{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b0880b14-77ff-4d7c-97b6-de495419611b",
   "metadata": {},
   "source": [
    "Resources for analysis:\n",
    "\n",
    "* https://ucdavis-bioinformatics-training.github.io/2022-June-RNA-Seq-Analysis/data_analysis/DE_Analysis_mm_with_quizzes_fixed\n",
    "* https://ucdavis-bioinformatics-training.github.io/2022-June-RNA-Seq-Analysis/data_reduction/03-counts_mm\n",
    "* https://github.com/ben-laufer/RNA-seq/blob/main/04-limma-voom.R\n",
    "* https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2020-mRNA_Seq_Workshop/master/data_analysis/enrichment_mm.Rmd\n",
    "* https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#selecting-an-ensembl-biomart-database-and-dataset\n",
    "* https://cran.r-project.org/web/packages/enrichR/vignettes/enrichR.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cda3d267-e5d4-48e8-95c3-e4cb82870a58",
   "metadata": {},
   "source": [
    "## Set Library Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4ce47088-d751-4c1d-90fd-f2153c7257d5",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/share/korflab/home/viki/anaconda3/jupyter_nb_R4.3/lib/R/library\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7682baae-5e1e-40cf-b4fe-20170dcc7a39",
   "metadata": {},
   "source": [
    "## Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c8478d6-9d45-4105-a316-f60650bef8c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(enrichR)\n",
    "library(dplyr)\n",
    "library(biomaRt)\n",
    "library(openxlsx)\n",
    "library(ggplot2)\n",
    "library(readxl)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1dfb0b5-c92e-4821-9b7f-0de10d5a631a",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2362d823-c1c4-4101-94c4-61448a02f5b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"05_DEGs/control_male_vs_control_female_results.csv\"   \n",
      "[2] \"05_DEGs/faexcess_female_vs_control_female_results.csv\"\n",
      "[3] \"05_DEGs/faexcess_female_vs_control_male_results.csv\"  \n",
      "[4] \"05_DEGs/faexcess_male_vs_control_female_results.csv\"  \n",
      "[5] \"05_DEGs/faexcess_male_vs_control_male_results.csv\"    \n",
      "[6] \"05_DEGs/faexcess_male_vs_faexcess_female_results.csv\" \n",
      "[7] \"05_DEGs/faexcess_vs_control_results.csv\"              \n"
     ]
    }
   ],
   "source": [
    "# List all DEG files\n",
    "files <- list.files(path = \"05_DEGs\", pattern = \"\\\\_results.csv$\", full.names = TRUE)\n",
    "\n",
    "# View\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "950300ee-5640-4c14-abf0-5387bbd91832",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select Ensembl database\n",
    "ensembl <- useEnsembl(biomart = \"genes\", dataset = \"mmusculus_gene_ensembl\", mirror = \"useast\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb0a6736-f3ff-4d7f-9fd9-909e93a07f73",
   "metadata": {},
   "source": [
    "## Convert to Gene Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3a39cfcd-341c-4a2d-a20b-07759350327c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processed and saved results for 05_DEGs/control_male_vs_control_female_results.csv to control_male_vs_control_female_genes.csv\n",
      "\n",
      "Processed and saved results for 05_DEGs/faexcess_female_vs_control_female_results.csv to faexcess_female_vs_control_female_genes.csv\n",
      "\n",
      "Processed and saved results for 05_DEGs/faexcess_female_vs_control_male_results.csv to faexcess_female_vs_control_male_genes.csv\n",
      "\n",
      "Processed and saved results for 05_DEGs/faexcess_male_vs_control_female_results.csv to faexcess_male_vs_control_female_genes.csv\n",
      "\n",
      "Processed and saved results for 05_DEGs/faexcess_male_vs_control_male_results.csv to faexcess_male_vs_control_male_genes.csv\n",
      "\n",
      "Processed and saved results for 05_DEGs/faexcess_male_vs_faexcess_female_results.csv to faexcess_male_vs_faexcess_female_genes.csv\n",
      "\n",
      "\n",
      "Batch submitting query [=======>-----------------------]  25% eta:  3m\n",
      "\n",
      "Batch submitting query [===============>---------------]  50% eta:  1m\n",
      "\n",
      "Batch submitting query [======================>--------]  75% eta: 24s\n",
      "                                                                      \n",
      "\n",
      "Processed and saved results for 05_DEGs/faexcess_vs_control_results.csv to faexcess_vs_control_genes.csv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Loop through each DEG file\n",
    "for (file in files) {\n",
    "  \n",
    "  # Read the current file\n",
    "  df <- read.csv(file)\n",
    "  \n",
    "  # Split the gene identifiers and extract gene names (assuming the gene IDs are in the \"X\" column)\n",
    "  split_genes <- strsplit(df$X, split = \"\\\\.\")\n",
    "  gene_names <- sapply(split_genes, \"[[\", 1L)\n",
    "  df$gene_names <- gene_names\n",
    "  \n",
    "  # Ensure geneList is a named vector for P.Values\n",
    "  geneList <- setNames(df$P.Value, df$gene_names)\n",
    "  \n",
    "  # Lookup Entrez Gene IDs for Ensembl Gene IDs (first lookup)\n",
    "  lookup_entrez <- getBM(\n",
    "    mart = ensembl,\n",
    "    attributes = c('ensembl_gene_id', 'entrezgene_id'),\n",
    "    filter = 'ensembl_gene_id',\n",
    "    values = df$gene_names,\n",
    "    uniqueRows = TRUE\n",
    "  )\n",
    "  \n",
    "  # Merge the lookup result for Entrez Gene IDs with the dataframe\n",
    "  df <- merge(df, lookup_entrez, by.x = \"gene_names\", by.y = \"ensembl_gene_id\", all.x = TRUE)\n",
    "  names(df)[names(df) == \"entrezgene_id\"] <- \"entrez_gene_id\"\n",
    "  \n",
    "  # Lookup external gene names for Ensembl Gene IDs (second lookup)\n",
    "  lookup_external <- getBM(\n",
    "    mart = ensembl,\n",
    "    attributes = c('ensembl_gene_id', 'external_gene_name'),\n",
    "    filter = 'ensembl_gene_id',\n",
    "    values = df$gene_names,\n",
    "    uniqueRows = TRUE\n",
    "  )\n",
    "  names(lookup_external)[names(lookup_external) == \"ensembl_gene_id\"] <- \"gene_names\"\n",
    "  \n",
    "  # Add external gene names to the dataframe\n",
    "  df <- df %>%\n",
    "    left_join(lookup_external, by = \"gene_names\") %>%\n",
    "    mutate(external_gene_name = ifelse(is.na(external_gene_name), \"NA\", external_gene_name))\n",
    "  \n",
    "  # Reorder columns for consistency\n",
    "  df <- df[, c(\"gene_names\", \"external_gene_name\", \"X\", \"logFC\", \"AveExpr\", \"t\", \"P.Value\", \"adj.P.Val\", \"B\", \"entrez_gene_id\")]\n",
    "  \n",
    "  # Filter to only unique rows (remove duplicate gene names)\n",
    "  df <- df %>%\n",
    "    filter(!is.na(external_gene_name) & external_gene_name != \"\") %>%\n",
    "    distinct(external_gene_name, .keep_all = TRUE)\n",
    "  \n",
    "  # Save the results to a new file\n",
    "  output_file <- sub(\"_results.csv$\", \"_genes.csv\", basename(file))\n",
    "  write.csv(df, file.path(dirname(file), output_file), row.names = FALSE)\n",
    "  \n",
    "  # Print a message indicating the completion for this file (optional)\n",
    "  message(paste(\"Processed and saved results for\", file, \"to\", output_file))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a1fd1e2-1404-4bc0-9e08-65d3b3e6bf09",
   "metadata": {},
   "source": [
    "## Gene Ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "364e4ee2-2c16-4bec-8a66-351a174b8670",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"05_DEGs/control_male_vs_control_female_genes.csv\"   \n",
      "[2] \"05_DEGs/faexcess_female_vs_control_female_genes.csv\"\n",
      "[3] \"05_DEGs/faexcess_female_vs_control_male_genes.csv\"  \n",
      "[4] \"05_DEGs/faexcess_male_vs_control_female_genes.csv\"  \n",
      "[5] \"05_DEGs/faexcess_male_vs_control_male_genes.csv\"    \n",
      "[6] \"05_DEGs/faexcess_male_vs_faexcess_female_genes.csv\" \n",
      "[7] \"05_DEGs/faexcess_vs_control_genes.csv\"              \n"
     ]
    }
   ],
   "source": [
    "# List all DEG files\n",
    "files <- list.files(path = \"05_DEGs\", pattern = \"\\\\_genes.csv$\", full.names = TRUE)\n",
    "\n",
    "# View\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "17231419-b05e-42f9-9490-49ef8065d1e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: control_male_vs_control_female_genes.csv - Number of significant DEGs: 0 \n",
      "File: faexcess_female_vs_control_female_genes.csv - Number of significant DEGs: 161 \n",
      "File: faexcess_female_vs_control_male_genes.csv - Number of significant DEGs: 0 \n",
      "File: faexcess_male_vs_control_female_genes.csv - Number of significant DEGs: 0 \n",
      "File: faexcess_male_vs_control_male_genes.csv - Number of significant DEGs: 0 \n",
      "File: faexcess_male_vs_faexcess_female_genes.csv - Number of significant DEGs: 0 \n",
      "File: faexcess_vs_control_genes.csv - Number of significant DEGs: 646 \n",
      "\n",
      "Files with significant DEGs:\n",
      "[1] \"05_DEGs/faexcess_female_vs_control_female_genes.csv\"\n",
      "[2] \"05_DEGs/faexcess_vs_control_genes.csv\"              \n"
     ]
    }
   ],
   "source": [
    "# Filter to significant DEGs only\n",
    "\n",
    "# Initialize a vector to store files with significant DEGs\n",
    "significant_DEGs_files <- c()\n",
    "\n",
    "# Loop through each file\n",
    "for (file in files) {\n",
    "  # Read the CSV file\n",
    "  data <- read.csv(file)\n",
    "  \n",
    "  # Filter for significant DEGs\n",
    "  significant_DEGs <- data[data$adj.P.Val < 0.05, ]\n",
    "  \n",
    "  # Print the file name and number of significant DEGs\n",
    "  cat(\"File:\", basename(file), \"- Number of significant DEGs:\", nrow(significant_DEGs), \"\\n\")\n",
    "  \n",
    "  # Check if there are significant DEGs\n",
    "  if (nrow(significant_DEGs) > 0) {\n",
    "    significant_DEGs_files <- c(significant_DEGs_files, file)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Overwrite to reflect only significant DEGs\n",
    "files <- significant_DEGs_files\n",
    "\n",
    "# View files with significant DEGs\n",
    "cat(\"\\nFiles with significant DEGs:\\n\")\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2dd384f3-28b0-48b4-8538-243b9d38e260",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "  Querying KEGG_2019_Mouse... Done.\n",
      "  Querying Panther_2016... Done.\n",
      "  Querying Reactome_2016... Done.\n",
      "  Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.\n",
      "Parsing results... Done.\n",
      "Enrichr results saved for: 05_DEGs/faexcess_female_vs_control_female_genes.csv -> 05_DEGs/faexcess_female_vs_control_female_genes_enrichr_results.xlsx \n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "  Querying KEGG_2019_Mouse... Done.\n",
      "  Querying Panther_2016... Done.\n",
      "  Querying Reactome_2016... Done.\n",
      "  Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.\n",
      "Parsing results... Done.\n",
      "Enrichr results saved for: 05_DEGs/faexcess_vs_control_genes.csv -> 05_DEGs/faexcess_vs_control_genes_enrichr_results.xlsx \n"
     ]
    }
   ],
   "source": [
    "# Loop through each file\n",
    "for (file in files) {\n",
    "  \n",
    "  # Read the current file\n",
    "  df <- read.csv(file)\n",
    "\n",
    "  # Filter to only significant DEGs\n",
    " df <- df[df$adj.P.Val < 0.05, ]\n",
    "    \n",
    "  # Perform the enrichR analysis on the gene list\n",
    "  enrichr_results <- enrichr(df$external_gene_name, c(\"GO_Biological_Process_2023\",\n",
    "                                                      \"GO_Cellular_Component_2023\",\n",
    "                                                      \"GO_Molecular_Function_2023\",\n",
    "                                                      \"KEGG_2019_Mouse\",\n",
    "                                                      \"Panther_2016\",\n",
    "                                                      \"Reactome_2016\",\n",
    "                                                      \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\"))\n",
    "  \n",
    "  # Create a new Excel workbook for the current file\n",
    "  wb <- createWorkbook()\n",
    "  \n",
    "  # Loop through each table in enrichr_results and save it as a separate sheet in the Excel workbook\n",
    "  for (i in seq_along(enrichr_results)) {\n",
    "    # Extract the data frame from the list\n",
    "    enrichr_df <- enrichr_results[[i]]\n",
    "    \n",
    "    # Define the original sheet name\n",
    "    original_sheet_name <- names(enrichr_results)[i]\n",
    "    \n",
    "    # Modify the sheet name if it's specifically \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\"\n",
    "    sheet_name <- if (original_sheet_name == \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\") {\n",
    "      \"RNAseq_DiseaseGene_DrugSigs_GEO\"\n",
    "    } else {\n",
    "      original_sheet_name\n",
    "    }\n",
    "    \n",
    "    # Add the data frame as a new sheet in the Excel workbook\n",
    "    addWorksheet(wb, sheet_name)\n",
    "    writeData(wb, sheet = sheet_name, x = enrichr_df)\n",
    "  }\n",
    "  \n",
    "  # Define the output Excel filename based on the input file name\n",
    "  output_filename <- paste0(\"05_DEGs/\", tools::file_path_sans_ext(basename(file)), \"_enrichr_results.xlsx\")\n",
    "  \n",
    "  # Save the Excel workbook for this file\n",
    "  saveWorkbook(wb, output_filename, overwrite = TRUE)\n",
    "  \n",
    "  # Print message indicating successful save\n",
    "  cat(\"Enrichr results saved for:\", file, \"->\", output_filename, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5fa0d5a-79c7-47de-8fc2-99422c3b8425",
   "metadata": {},
   "source": [
    "## Plot All Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fb598162-662f-4670-ac5b-784a42e34d3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"05_DEGs/faexcess_female_vs_control_female_genes_enrichr_results.xlsx\"\n",
      "[2] \"05_DEGs/faexcess_vs_control_genes_enrichr_results.xlsx\"              \n"
     ]
    }
   ],
   "source": [
    "# List all Enrichr files\n",
    "files <- list.files(path = \"05_DEGs\", pattern = \"\\\\.xlsx$\", full.names = TRUE)\n",
    "\n",
    "# View\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3ee0c315-c00c-424a-9a29-2831efcce6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define functions to calculate -log10 of P-value and filter top 10 terms\n",
    "calculate_neg_log_p <- function(df) {\n",
    "  df %>%\n",
    "    mutate(neg_log_p = ifelse(P.value > 0, -log10(P.value), Inf))\n",
    "}\n",
    "\n",
    "filter_top_10 <- function(df) {\n",
    "  return(df %>% arrange(P.value) %>% head(10))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9d9c47bd-b54e-432e-be7b-e76da7cb8556",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating plot for faexcess_female vs control_female ...\n",
      "Plot generated for FAE Female vs Control Female \n",
      "Generating plot for faexcess vs control ...\n",
      "Plot generated for FAE vs Control \n"
     ]
    }
   ],
   "source": [
    "# Loop through each Excel file and make graphs of GO databases\n",
    "for (file in files) {\n",
    "  # Extract the base filename without extension\n",
    "  filename <- tools::file_path_sans_ext(basename(file))\n",
    "  \n",
    "  # Split the filename into parts based on '_vs_'\n",
    "  split_name <- unlist(strsplit(filename, \"_vs_\"))\n",
    "  \n",
    "  # Handle different naming conventions and assign group1 and group2\n",
    "  if (length(split_name) == 2) {\n",
    "    group1 <- tolower(split_name[1])  \n",
    "    group2 <- tolower(split_name[2])\n",
    "    # Remove \"_genes\" from group2 \n",
    "    group2 <- gsub(\"_genes_enrichr_results\", \"\", group2)\n",
    "    # Adjust group names based on specific cases\n",
    "    if (grepl(\"faexcess_female\", group1)) {\n",
    "      group1 <- \"faexcess_female\"\n",
    "    } else if (grepl(\"control_female\", group2)) {\n",
    "      group2 <- \"control_female\"\n",
    "    }\n",
    "  } else if (length(split_name) == 1) {\n",
    "    group1 <- tolower(split_name[1])\n",
    "    group2 <- \"control\"\n",
    "  }\n",
    "\n",
    "  # Confirm grouping\n",
    "  cat(\"Generating plot for\", group1, \"vs\", group2, \"...\\n\")\n",
    " \n",
    "  # Read Enrichr results from Excel sheets\n",
    "  enrichr_results <- list(\n",
    "    GO_Biological_Process_2023 = read_excel(file, sheet = \"GO_Biological_Process_2023\"),\n",
    "    GO_Cellular_Component_2023 = read_excel(file, sheet = \"GO_Cellular_Component_2023\"),\n",
    "    GO_Molecular_Function_2023 = read_excel(file, sheet = \"GO_Molecular_Function_2023\")\n",
    "  )\n",
    "    \n",
    "  # Plot Enrichr results\n",
    "  GO_BP <- enrichr_results$GO_Biological_Process_2023\n",
    "  GO_CC <- enrichr_results$GO_Cellular_Component_2023\n",
    "  GO_MF <- enrichr_results$GO_Molecular_Function_2023\n",
    "  \n",
    "  # Apply functions to each data frame\n",
    "  GO_BP <- GO_BP %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "  \n",
    "  GO_CC <- GO_CC %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "  \n",
    "  GO_MF <- GO_MF %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "\n",
    "  # Combine data frames for plotting\n",
    "  combined_df <- rbind(mutate(GO_BP, Category = \"Biological Process\"),\n",
    "                       mutate(GO_CC, Category = \"Cellular Component\"),\n",
    "                       mutate(GO_MF, Category = \"Molecular Function\"))\n",
    "    \n",
    "  # Filter out rows with NA in neg_log_p\n",
    "  combined_df <- combined_df %>% filter(!is.na(neg_log_p))\n",
    "\n",
    "  # Calculate the maximum limit for the x-axis\n",
    "  x_max <- ceiling(max(combined_df$neg_log_p, na.rm = TRUE))\n",
    "    \n",
    "  # Plot\n",
    "  pdf(paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_enrichr_results_barplot_all.pdf\"), height = 5, width = 12)\n",
    "\n",
    "  # Rename groups for plotting  \n",
    "  group1 <- ifelse(group1 == \"faexcess_female\", \"FAE Female\", \"FAE\")\n",
    "  group2 <- ifelse(group2 == \"control_female\", \"Control Female\", \"Control\")\n",
    "    \n",
    "  p <- ggplot(combined_df, aes(x = neg_log_p, y = reorder(Term, neg_log_p), fill = Category)) +\n",
    "    geom_col(position = position_dodge(width = 0.9)) +\n",
    "    scale_fill_manual(values = c(\"Biological Process\" = \"#D1D2F9\", \n",
    "                                  \"Cellular Component\" = \"#8FD3FE\", \n",
    "                                  \"Molecular Function\" = \"#BEE3BA\")) +\n",
    "    labs(x = \"-log10(P-value)\", y = NULL, title = paste(\"Enrichr Results for\", group1, \"vs.\", group2, \"for All DEGs\")) +\n",
    "    scale_x_continuous(limits = c(0, x_max)) +\n",
    "    facet_grid(Category ~ ., scales = \"free_y\", space = \"free_y\") +\n",
    "    theme_minimal() +\n",
    "    theme(legend.position = \"right\",\n",
    "          strip.text.y = element_blank(),  # Remove y-axis labels\n",
    "          panel.grid.major = element_blank(),  # Remove major grid lines\n",
    "          panel.grid.minor = element_blank())  # Remove minor grid lines\n",
    "\n",
    "  print(p)\n",
    "    \n",
    "  dev.off()\n",
    "\n",
    "  # Print progress\n",
    "  cat(\"Plot generated for\", group1, \"vs\", group2, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d0476f9e-bb7f-4027-9bdc-6dd7b952e773",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_GO_Biological_Process_2023_all.pdf \n",
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_GO_Cellular_Component_2023_all.pdf \n",
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_GO_Molecular_Function_2023_all.pdf \n",
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_KEGG_2019_Mouse_all.pdf \n",
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_Panther_2016_all.pdf \n",
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_Reactome_2016_all.pdf \n",
      "Plot saved for: faexcess_female vs control_female -> 05_DEGs/faexcess_female_vs_control_female_RNAseq_DiseaseGene_DrugSigs_GEO_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_GO_Biological_Process_2023_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_GO_Cellular_Component_2023_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_GO_Molecular_Function_2023_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_KEGG_2019_Mouse_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_Panther_2016_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_Reactome_2016_all.pdf \n",
      "Plot saved for: faexcess vs control -> 05_DEGs/faexcess_vs_control_RNAseq_DiseaseGene_DrugSigs_GEO_all.pdf \n"
     ]
    }
   ],
   "source": [
    "# Mapping of original sheet names to database names\n",
    "sheet_mapping <- c(\n",
    "  \"GO_Biological_Process_2023\" = \"GO_Biological_Process_2023\",\n",
    "  \"GO_Cellular_Component_2023\" = \"GO_Cellular_Component_2023\",\n",
    "  \"GO_Molecular_Function_2023\" = \"GO_Molecular_Function_2023\",\n",
    "  \"KEGG_2019_Mouse\" = \"KEGG_2019_Mouse\",\n",
    "  \"Panther_2016\" = \"Panther_2016\",\n",
    "  \"Reactome_2016\" = \"Reactome_2016\",\n",
    "  \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\" = \"RNAseq_DiseaseGene_DrugSigs_GEO\"\n",
    ")\n",
    "\n",
    "# Loop through each Excel file and plot Enrichr results per database\n",
    "for (file in files) {\n",
    "  # Extract the base filename without extension\n",
    "  filename <- tools::file_path_sans_ext(basename(file))\n",
    "  \n",
    "  # Split the filename into parts based on '_vs_'\n",
    "  split_name <- unlist(strsplit(filename, \"_vs_\"))\n",
    "  \n",
    "  # Handle different naming conventions and assign group1 and group2\n",
    "  if (length(split_name) == 2) {\n",
    "    # Case for filename \"control_female_vs_faexcess_female\"\n",
    "    group1 <- tolower(split_name[1])  \n",
    "    group2 <- tolower(gsub(\"_genes_enrichr_results\", \"\", split_name[2]))  \n",
    "  } else if (length(split_name) == 1) {\n",
    "    # Case for filename \"control_vs_faexcess\"\n",
    "    group1 <- tolower(split_name[1])\n",
    "    group2 <- \"faexcess\"\n",
    "  }\n",
    "  \n",
    "  # Initialize a flag to track if valid data was found and processed\n",
    "  valid_data_found <- FALSE\n",
    "  \n",
    "  # List of GO term sheets\n",
    "  sheets <- names(sheet_mapping)  # Use the names of the mapping as sheets\n",
    "  \n",
    "  # Loop through each sheet and process\n",
    "  for (sheet in sheets) {\n",
    "    tryCatch({\n",
    "      # Read the sheet using the mapping\n",
    "      GO_terms <- read_excel(file, sheet = sheet_mapping[sheet])  # Read each sheet\n",
    "\n",
    "      # Only proceed if data is non-null and has rows\n",
    "      if (!is.null(GO_terms) && nrow(GO_terms) > 0) {\n",
    "        # Mark that valid data was found\n",
    "        valid_data_found <- TRUE\n",
    "        \n",
    "        # Define the PDF filename\n",
    "        pdf_filename <- paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_\", sheet_mapping[sheet], \"_all.pdf\")\n",
    "        \n",
    "        # Create the plot and save it\n",
    "        pdf(pdf_filename, height = 7, width = 15)\n",
    "        p <- plotEnrich(GO_terms, showTerms = 25, numChar = 75, y = \"Count\", orderBy = \"P.value\") +\n",
    "          ggtitle(paste(sheet_mapping[sheet], \"for all DEGs in\", group1, \"vs.\", group2))\n",
    "        \n",
    "        print(p)\n",
    "        dev.off()\n",
    "        \n",
    "        # Print a message indicating the plot was saved\n",
    "        cat(\"Plot saved for:\", group1, \"vs\", group2, \"->\", pdf_filename, \"\\n\")\n",
    "      } else {\n",
    "        # If no valid data is found, skip plotting for this sheet\n",
    "        cat(\"No data found for\", sheet, \"in:\", file, \"\\n\")\n",
    "      }\n",
    "      \n",
    "    }, error = function(e) {\n",
    "      # Handle any read errors\n",
    "      cat(\"Error reading\", sheet, \"in:\", file, \"\\n\")\n",
    "    })\n",
    "  }\n",
    "  \n",
    "  # If no valid data was found for the entire file, print that it was skipped\n",
    "  if (!valid_data_found) {\n",
    "    cat(\"No data found for:\", group1, \"vs\", group2, \". Skipping plots.\\n\")\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6bd330-f95d-417e-85f2-f9405dbbccf9",
   "metadata": {},
   "source": [
    "## Plot Upregulated Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "83098f36-d889-45ac-a735-69ddf40fca4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: faexcess_female_vs_control_female_genes.csv - Number of significant DEGs: 161 \n",
      "File: faexcess_vs_control_genes.csv - Number of significant DEGs: 646 \n",
      "\n",
      "Files with significant DEGs:\n",
      "[1] \"05_DEGs/faexcess_female_vs_control_female_genes.csv\"\n",
      "[2] \"05_DEGs/faexcess_vs_control_genes.csv\"              \n"
     ]
    }
   ],
   "source": [
    "# List of result files\n",
    "files <- list.files(\"05_DEGs\", pattern = \"_genes\\\\.csv$\", full.names = TRUE)\n",
    "\n",
    "# Filter to significant DEGs only\n",
    "significant_DEGs_files <- c()\n",
    "\n",
    "# Loop through each file\n",
    "for (file in files) {\n",
    "  # Read the CSV file\n",
    "  data <- read.csv(file)\n",
    "  \n",
    "  # Filter for significant DEGs\n",
    "  significant_DEGs <- data[data$adj.P.Val < 0.05, ]\n",
    "  \n",
    "  # Print the file name and number of significant DEGs\n",
    "  cat(\"File:\", basename(file), \"- Number of significant DEGs:\", nrow(significant_DEGs), \"\\n\")\n",
    "  \n",
    "  # Check if there are significant DEGs\n",
    "  if (nrow(significant_DEGs) > 0) {\n",
    "    significant_DEGs_files <- c(significant_DEGs_files, file)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Overwrite to reflect only significant DEGs\n",
    "files <- significant_DEGs_files\n",
    "\n",
    "# View files with significant DEGs\n",
    "cat(\"\\nFiles with significant DEGs:\\n\")\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "57f9ad7b-6176-4361-91be-33b1d17bc94f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating plot for faexcess_female vs control_female ...\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "Parsing results... Done.\n",
      "Plot generated for FAE Female vs Control Female \n",
      "Generating plot for faexcess vs control ...\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "Parsing results... Done.\n",
      "Plot generated for FAE vs Control \n"
     ]
    }
   ],
   "source": [
    "# Loop through each result file\n",
    "for (file in files) {\n",
    "  # Extract the base filename without extension\n",
    "  filename <- tools::file_path_sans_ext(basename(file))\n",
    "  \n",
    "  # Split the filename into parts based on '_vs_'\n",
    "  split_name <- unlist(strsplit(filename, \"_vs_\"))\n",
    "  \n",
    "  # Handle different naming conventions and assign group1 and group2\n",
    "  if (length(split_name) == 2) {\n",
    "    group1 <- tolower(split_name[1])  \n",
    "    group2 <- tolower(split_name[2])\n",
    "    # Remove \"_genes\" from group2 \n",
    "    group2 <- gsub(\"_genes\", \"\", group2)\n",
    "    # Adjust group names based on specific cases\n",
    "    if (grepl(\"faexcess_female\", group1)) {\n",
    "      group1 <- \"faexcess_female\"\n",
    "    } else if (grepl(\"control_female\", group2)) {\n",
    "      group2 <- \"control_female\"\n",
    "    }\n",
    "  } else if (length(split_name) == 1) {\n",
    "    group1 <- tolower(split_name[1])\n",
    "    group2 <- \"control\"\n",
    "  }\n",
    "\n",
    "  # Confirm grouping\n",
    "  cat(\"Generating plot for\", group1, \"vs\", group2, \"...\\n\")\n",
    "    \n",
    "  # Read in data\n",
    "  df <- read.csv(file)\n",
    "    \n",
    "  # Filter to only significant DEGs\n",
    "  df <- df[df$adj.P.Val < 0.05, ]\n",
    "  \n",
    "  # Filter for only upregulated genes\n",
    "  tmp <- df[df$logFC > 0, ]\n",
    "  \n",
    "  # Check if there are any upregulated genes\n",
    "  if (nrow(tmp) == 0) {\n",
    "    cat(\"No significant upregulated genes found for:\", group1, \"vs\", group2, \"\\n\")\n",
    "    next\n",
    "  }\n",
    "  \n",
    "  # Perform the enrichR analysis on the gene list\n",
    "  enrichr_results <- enrichr(tmp$external_gene_name, c(\"GO_Biological_Process_2023\",\n",
    "                                                       \"GO_Cellular_Component_2023\",\n",
    "                                                       \"GO_Molecular_Function_2023\"))\n",
    "    \n",
    "  # Plot Enrichr results\n",
    "  GO_BP <- enrichr_results$GO_Biological_Process_2023\n",
    "  GO_CC <- enrichr_results$GO_Cellular_Component_2023\n",
    "  GO_MF <- enrichr_results$GO_Molecular_Function_2023\n",
    "  \n",
    "  # Apply functions to each data frame\n",
    "  GO_BP <- GO_BP %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "  \n",
    "  GO_CC <- GO_CC %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "  \n",
    "  GO_MF <- GO_MF %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "\n",
    "  # Combine data frames for plotting\n",
    "  combined_df <- rbind(mutate(GO_BP, Category = \"Biological Process\"),\n",
    "                       mutate(GO_CC, Category = \"Cellular Component\"),\n",
    "                       mutate(GO_MF, Category = \"Molecular Function\"))\n",
    "    \n",
    "  # Filter out rows with NA in neg_log_p\n",
    "  combined_df <- combined_df %>% filter(!is.na(neg_log_p))\n",
    "\n",
    "  # Calculate the maximum limit for the x-axis\n",
    "  x_max <- ceiling(max(combined_df$neg_log_p, na.rm = TRUE))\n",
    "    \n",
    "  # Plot\n",
    "  pdf(paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_enrichr_results_barplot_up.pdf\"), height = 5, width = 12)\n",
    "\n",
    "  # Rename groups for plotting  \n",
    "  group1 <- ifelse(group1 == \"faexcess_female\", \"FAE Female\", \"FAE\")\n",
    "  group2 <- ifelse(group2 == \"control_female\", \"Control Female\", \"Control\")\n",
    "    \n",
    "  p <- ggplot(combined_df, aes(x = neg_log_p, y = reorder(Term, neg_log_p), fill = Category)) +\n",
    "    geom_col(position = position_dodge(width = 0.9)) +\n",
    "    scale_fill_manual(values = c(\"Biological Process\" = \"#D1D2F9\", \n",
    "                                  \"Cellular Component\" = \"#8FD3FE\", \n",
    "                                  \"Molecular Function\" = \"#BEE3BA\")) +\n",
    "    labs(x = \"-log10(P-value)\", y = NULL, title = paste(\"Enrichr Results for\", group1, \"vs.\", group2, \"for Upregulated DEGs\")) +\n",
    "    scale_x_continuous(limits = c(0, x_max)) +\n",
    "    facet_grid(Category ~ ., scales = \"free_y\", space = \"free_y\") +\n",
    "    theme_minimal() +\n",
    "    theme(legend.position = \"right\",\n",
    "          strip.text.y = element_blank(),  # Remove y-axis labels\n",
    "          panel.grid.major = element_blank(),  # Remove major grid lines\n",
    "          panel.grid.minor = element_blank())  # Remove minor grid lines\n",
    "\n",
    "  print(p)\n",
    "    \n",
    "  dev.off()\n",
    "\n",
    "  # Print progress\n",
    "  cat(\"Plot generated for\", group1, \"vs\", group2, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a399c543-e8a5-4c06-86a0-606eebc71862",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "  Querying KEGG_2019_Mouse... Done.\n",
      "  Querying Panther_2016... Done.\n",
      "  Querying Reactome_2016... Done.\n",
      "  Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.\n",
      "Parsing results... Done.\n",
      "Plot saved for: GO_Biological_Process_2023 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_GO_Biological_Process_2023_up.pdf \n",
      "Plot saved for: GO_Cellular_Component_2023 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_GO_Cellular_Component_2023_up.pdf \n",
      "Plot saved for: GO_Molecular_Function_2023 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_GO_Molecular_Function_2023_up.pdf \n",
      "Plot saved for: KEGG_2021_Human for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_KEGG_2019_Mouse_up.pdf \n",
      "Plot saved for: Panther_2016 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_Panther_2016_up.pdf \n",
      "Plot saved for: Reactome_2016 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_Reactome_2016_up.pdf \n",
      "Plot saved for: RNA_Seq_Disease_Gene_and_Drug_Signatures_from_GEO for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_up.pdf \n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "  Querying KEGG_2019_Mouse... Done.\n",
      "  Querying Panther_2016... Done.\n",
      "  Querying Reactome_2016... Done.\n",
      "  Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.\n",
      "Parsing results... Done.\n",
      "Plot saved for: GO_Biological_Process_2023 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_GO_Biological_Process_2023_up.pdf \n",
      "Plot saved for: GO_Cellular_Component_2023 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_GO_Cellular_Component_2023_up.pdf \n",
      "Plot saved for: GO_Molecular_Function_2023 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_GO_Molecular_Function_2023_up.pdf \n",
      "Plot saved for: KEGG_2021_Human for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_KEGG_2019_Mouse_up.pdf \n",
      "Plot saved for: Panther_2016 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_Panther_2016_up.pdf \n",
      "Plot saved for: Reactome_2016 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_Reactome_2016_up.pdf \n",
      "Plot saved for: RNA_Seq_Disease_Gene_and_Drug_Signatures_from_GEO for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_up.pdf \n"
     ]
    }
   ],
   "source": [
    "# Loop through each result file\n",
    "for (file in files) {\n",
    "  # Extract the base filename without extension\n",
    "  filename <- tools::file_path_sans_ext(basename(file))\n",
    "  \n",
    "  # Split the filename into parts based on '_vs_'\n",
    "  split_name <- unlist(strsplit(filename, \"_vs_\"))\n",
    "  \n",
    "  # Handle different naming conventions and assign group1 and group2\n",
    "  if (length(split_name) == 2) {\n",
    "    # Case for filename \"control_female_vs_faexcess_female\"\n",
    "    group1 <- tolower(split_name[1])  \n",
    "    group2 <- tolower(gsub(\"_genes_enrichr_results\", \"\", split_name[2]))  \n",
    "  } else if (length(split_name) == 1) {\n",
    "    # Case for filename \"control_vs_faexcess\"\n",
    "    group1 <- tolower(split_name[1])\n",
    "    group2 <- \"faexcess\"\n",
    "  }\n",
    "\n",
    "  # Read in data\n",
    "  df <- read.csv(file)\n",
    "    \n",
    "  # Filter to only significant DEGs\n",
    "  df <- df[df$adj.P.Val < 0.05, ]\n",
    "  \n",
    "  # Filter for only upregulated genes\n",
    "  tmp <- df[df$logFC > 0, ]\n",
    "  \n",
    "  # Check if there are any upregulated genes\n",
    "  if (nrow(tmp) == 0) {\n",
    "    cat(\"No significant upregulated genes found for:\", group1, \"vs\", group2, \"\\n\")\n",
    "    next\n",
    "  }\n",
    "  \n",
    "  # Perform the enrichR analysis on the gene list\n",
    "  enrichr_results <- enrichr(tmp$external_gene_name, c(\"GO_Biological_Process_2023\",\n",
    "                                                       \"GO_Cellular_Component_2023\",\n",
    "                                                       \"GO_Molecular_Function_2023\",\n",
    "                                                       \"KEGG_2019_Mouse\",\n",
    "                                                       \"Panther_2016\",\n",
    "                                                       \"Reactome_2016\",\n",
    "                                                       \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\"))\n",
    "\n",
    "  # Define the output file paths for each enrichment plot\n",
    "  output_files <- list(\n",
    "    GO_Biological_Process_2023 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_GO_Biological_Process_2023_up.pdf\"),\n",
    "    GO_Cellular_Component_2023 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_GO_Cellular_Component_2023_up.pdf\"),\n",
    "    GO_Molecular_Function_2023 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_GO_Molecular_Function_2023_up.pdf\"),\n",
    "    KEGG_2021_Human = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_KEGG_2019_Mouse_up.pdf\"),\n",
    "    Panther_2016 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_Panther_2016_up.pdf\"),\n",
    "    Reactome_2016 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_Reactome_2016_up.pdf\"),\n",
    "    RNA_Seq_Disease_Gene_and_Drug_Signatures_from_GEO = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_up.pdf\")\n",
    "  )\n",
    "\n",
    "  # Loop through each output file and plot the results\n",
    "  for (i in seq_along(enrichr_results)) {\n",
    "    if (!is.null(enrichr_results[[i]]) && nrow(enrichr_results[[i]]) > 0) {\n",
    "      pdf(output_files[[i]], height = 7, width = 15)\n",
    "      p <- plotEnrich(enrichr_results[[i]], showTerms = 25, numChar = 75, y = \"Count\", orderBy = \"P.value\") +\n",
    "        ggtitle(paste(names(output_files)[i], \"for DEGs Upregulated in\", group1, \"vs\", group2))\n",
    "      print(p)\n",
    "      dev.off()\n",
    "      \n",
    "      # Print message indicating the plot was saved\n",
    "      cat(\"Plot saved for:\", names(output_files)[i], \"for\", group1, \"vs\", group2, \"->\", output_files[[i]], \"\\n\")\n",
    "    } else {\n",
    "      # Print a message indicating that there were no results to plot\n",
    "      cat(\"No results to plot for:\", names(output_files)[i], \"for\", group1, \"vs\", group2, \"\\n\")\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9374edb-a2c5-4517-baf9-7f1aa6af8d72",
   "metadata": {},
   "source": [
    "## Plot Downregulated Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f75aaa57-a6ab-4e67-bfbb-53e0fa4ec02a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: faexcess_female_vs_control_female_genes.csv - Number of significant DEGs: 161 \n",
      "File: faexcess_vs_control_genes.csv - Number of significant DEGs: 646 \n",
      "\n",
      "Files with significant DEGs:\n",
      "[1] \"05_DEGs/faexcess_female_vs_control_female_genes.csv\"\n",
      "[2] \"05_DEGs/faexcess_vs_control_genes.csv\"              \n"
     ]
    }
   ],
   "source": [
    "# List of result files\n",
    "files <- list.files(\"05_DEGs\", pattern = \"_genes\\\\.csv$\", full.names = TRUE)\n",
    "\n",
    "# Filter to significant DEGs only\n",
    "significant_DEGs_files <- c()\n",
    "\n",
    "# Loop through each file\n",
    "for (file in files) {\n",
    "  # Read the CSV file\n",
    "  data <- read.csv(file)\n",
    "  \n",
    "  # Filter for significant DEGs\n",
    "  significant_DEGs <- data[data$adj.P.Val < 0.05, ]\n",
    "  \n",
    "  # Print the file name and number of significant DEGs\n",
    "  cat(\"File:\", basename(file), \"- Number of significant DEGs:\", nrow(significant_DEGs), \"\\n\")\n",
    "  \n",
    "  # Check if there are significant DEGs\n",
    "  if (nrow(significant_DEGs) > 0) {\n",
    "    significant_DEGs_files <- c(significant_DEGs_files, file)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Overwrite to reflect only significant DEGs\n",
    "files <- significant_DEGs_files\n",
    "\n",
    "# View files with significant DEGs\n",
    "cat(\"\\nFiles with significant DEGs:\\n\")\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "38108166-3a94-4fa1-9607-f1b17ac27e34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating plot for faexcess_female vs control_female ...\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "Parsing results... Done.\n",
      "Plot generated for FAE Female vs Control Female \n",
      "Generating plot for faexcess vs control ...\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "Parsing results... Done.\n",
      "Plot generated for FAE vs Control \n"
     ]
    }
   ],
   "source": [
    "# Loop through each result file\n",
    "for (file in files) {\n",
    "  # Extract the base filename without extension\n",
    "  filename <- tools::file_path_sans_ext(basename(file))\n",
    "  \n",
    "  # Split the filename into parts based on '_vs_'\n",
    "  split_name <- unlist(strsplit(filename, \"_vs_\"))\n",
    "  \n",
    "  # Handle different naming conventions and assign group1 and group2\n",
    "  if (length(split_name) == 2) {\n",
    "    group1 <- tolower(split_name[1])  \n",
    "    group2 <- tolower(split_name[2])\n",
    "    # Remove \"_genes\" from group2 \n",
    "    group2 <- gsub(\"_genes\", \"\", group2)\n",
    "    # Adjust group names based on specific cases\n",
    "    if (grepl(\"faexcess_female\", group1)) {\n",
    "      group1 <- \"faexcess_female\"\n",
    "    } else if (grepl(\"control_female\", group2)) {\n",
    "      group2 <- \"control_female\"\n",
    "    }\n",
    "  } else if (length(split_name) == 1) {\n",
    "    group1 <- tolower(split_name[1])\n",
    "    group2 <- \"control\"\n",
    "  }\n",
    "\n",
    "  # Confirm grouping\n",
    "  cat(\"Generating plot for\", group1, \"vs\", group2, \"...\\n\")\n",
    "    \n",
    "  # Read in data\n",
    "  df <- read.csv(file)\n",
    "    \n",
    "  # Filter to only significant DEGs\n",
    "  df <- df[df$adj.P.Val < 0.05, ]\n",
    "  \n",
    "  # Filter for only downregulated genes\n",
    "  tmp <- df[df$logFC < 0, ]\n",
    "  \n",
    "  # Check if there are any downregulated genes\n",
    "  if (nrow(tmp) == 0) {\n",
    "    cat(\"No significant downregulated genes found for:\", group1, \"vs\", group2, \"\\n\")\n",
    "    next\n",
    "  }\n",
    "  \n",
    "  # Perform the enrichR analysis on the gene list\n",
    "  enrichr_results <- enrichr(tmp$external_gene_name, c(\"GO_Biological_Process_2023\",\n",
    "                                                       \"GO_Cellular_Component_2023\",\n",
    "                                                       \"GO_Molecular_Function_2023\"))\n",
    "    \n",
    "  # Plot Enrichr results\n",
    "  GO_BP <- enrichr_results$GO_Biological_Process_2023\n",
    "  GO_CC <- enrichr_results$GO_Cellular_Component_2023\n",
    "  GO_MF <- enrichr_results$GO_Molecular_Function_2023\n",
    "  \n",
    "  # Apply functions to each data frame\n",
    "  GO_BP <- GO_BP %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "  \n",
    "  GO_CC <- GO_CC %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "  \n",
    "  GO_MF <- GO_MF %>%\n",
    "    filter_top_10() %>%\n",
    "    calculate_neg_log_p()\n",
    "\n",
    "  # Combine data frames for plotting\n",
    "  combined_df <- rbind(mutate(GO_BP, Category = \"Biological Process\"),\n",
    "                       mutate(GO_CC, Category = \"Cellular Component\"),\n",
    "                       mutate(GO_MF, Category = \"Molecular Function\"))\n",
    "    \n",
    "  # Filter out rows with NA in neg_log_p\n",
    "  combined_df <- combined_df %>% filter(!is.na(neg_log_p))\n",
    "\n",
    "  # Calculate the maximum limit for the x-axis\n",
    "  x_max <- ceiling(max(combined_df$neg_log_p, na.rm = TRUE))\n",
    "    \n",
    "  # Plot\n",
    "  pdf(paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_enrichr_results_barplot_down.pdf\"), height = 5, width = 12)\n",
    "\n",
    "  # Rename groups for plotting  \n",
    "  group1 <- ifelse(group1 == \"faexcess_female\", \"FAE Female\", \"FAE\")\n",
    "  group2 <- ifelse(group2 == \"control_female\", \"Control Female\", \"Control\")\n",
    "    \n",
    "  p <- ggplot(combined_df, aes(x = neg_log_p, y = reorder(Term, neg_log_p), fill = Category)) +\n",
    "    geom_col(position = position_dodge(width = 0.9)) +\n",
    "    scale_fill_manual(values = c(\"Biological Process\" = \"#D1D2F9\", \n",
    "                                  \"Cellular Component\" = \"#8FD3FE\", \n",
    "                                  \"Molecular Function\" = \"#BEE3BA\")) +\n",
    "    labs(x = \"-log10(P-value)\", y = NULL, title = paste(\"Enrichr Results for\", group1, \"vs.\", group2, \"for Downregulated DEGs\")) +\n",
    "    scale_x_continuous(limits = c(0, x_max)) +\n",
    "    facet_grid(Category ~ ., scales = \"free_y\", space = \"free_y\") +\n",
    "    theme_minimal() +\n",
    "    theme(legend.position = \"right\",\n",
    "          strip.text.y = element_blank(),  # Remove y-axis labels\n",
    "          panel.grid.major = element_blank(),  # Remove major grid lines\n",
    "          panel.grid.minor = element_blank())  # Remove minor grid lines\n",
    "\n",
    "  print(p)\n",
    "    \n",
    "  dev.off()\n",
    "\n",
    "  # Print progress\n",
    "  cat(\"Plot generated for\", group1, \"vs\", group2, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9c8c28fb-3d9a-4283-9755-a8c4572b5c8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "  Querying KEGG_2019_Mouse... Done.\n",
      "  Querying Panther_2016... Done.\n",
      "  Querying Reactome_2016... Done.\n",
      "  Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.\n",
      "Parsing results... Done.\n",
      "Enrichment results for: GO_Biological_Process_2023 - Rows: 368 \n",
      "Enrichment results for: GO_Cellular_Component_2023 - Rows: 93 \n",
      "Enrichment results for: GO_Molecular_Function_2023 - Rows: 101 \n",
      "Enrichment results for: KEGG_2019_Mouse - Rows: 41 \n",
      "Enrichment results for: Panther_2016 - Rows: 4 \n",
      "Enrichment results for: Reactome_2016 - Rows: 460 \n",
      "Enrichment results for: RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO - Rows: 1162 \n",
      "Plot saved for: GO_Biological_Process_2023 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_GO_Biological_Process_2023_down.pdf \n",
      "Plot saved for: GO_Cellular_Component_2023 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_GO_Cellular_Component_2023_down.pdf \n",
      "Plot saved for: GO_Molecular_Function_2023 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_GO_Molecular_Function_2023_down.pdf \n",
      "Plot saved for: KEGG_2021_Human for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_KEGG_2019_Mouse_down.pdf \n",
      "Plot saved for: Panther_2016 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_Panther_2016_down.pdf \n",
      "Plot saved for: Reactome_2016 for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_Reactome_2016_down.pdf \n",
      "Plot saved for: RNA_Seq_Disease_Gene_and_Drug_Signatures_from_GEO for faexcess_female vs control_female_genes -> 05_DEGs/faexcess_female_vs_control_female_genes_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_down.pdf \n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2023... Done.\n",
      "  Querying GO_Cellular_Component_2023... Done.\n",
      "  Querying GO_Molecular_Function_2023... Done.\n",
      "  Querying KEGG_2019_Mouse... Done.\n",
      "  Querying Panther_2016... Done.\n",
      "  Querying Reactome_2016... Done.\n",
      "  Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.\n",
      "Parsing results... Done.\n",
      "Enrichment results for: GO_Biological_Process_2023 - Rows: 1345 \n",
      "Enrichment results for: GO_Cellular_Component_2023 - Rows: 182 \n",
      "Enrichment results for: GO_Molecular_Function_2023 - Rows: 280 \n",
      "Enrichment results for: KEGG_2019_Mouse - Rows: 164 \n",
      "Enrichment results for: Panther_2016 - Rows: 25 \n",
      "Enrichment results for: Reactome_2016 - Rows: 665 \n",
      "Enrichment results for: RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO - Rows: 1281 \n",
      "Plot saved for: GO_Biological_Process_2023 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_GO_Biological_Process_2023_down.pdf \n",
      "Plot saved for: GO_Cellular_Component_2023 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_GO_Cellular_Component_2023_down.pdf \n",
      "Plot saved for: GO_Molecular_Function_2023 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_GO_Molecular_Function_2023_down.pdf \n",
      "Plot saved for: KEGG_2021_Human for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_KEGG_2019_Mouse_down.pdf \n",
      "Plot saved for: Panther_2016 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_Panther_2016_down.pdf \n",
      "Plot saved for: Reactome_2016 for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_Reactome_2016_down.pdf \n",
      "Plot saved for: RNA_Seq_Disease_Gene_and_Drug_Signatures_from_GEO for faexcess vs control_genes -> 05_DEGs/faexcess_vs_control_genes_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_down.pdf \n"
     ]
    }
   ],
   "source": [
    "# Loop through each result file\n",
    "for (file in files) {\n",
    "  # Extract the base filename without extension\n",
    "  filename <- tools::file_path_sans_ext(basename(file))\n",
    "  \n",
    "  # Split the filename into parts based on '_vs_'\n",
    "  split_name <- unlist(strsplit(filename, \"_vs_\"))\n",
    "  \n",
    "  # Handle different naming conventions and assign group1 and group2\n",
    "  if (length(split_name) == 2) {\n",
    "    # Case for filename \"control_female_vs_faexcess_female\"\n",
    "    group1 <- tolower(split_name[1])  \n",
    "    group2 <- tolower(gsub(\"_genes_enrichr_results\", \"\", split_name[2]))  \n",
    "  } else if (length(split_name) == 1) {\n",
    "    # Case for filename \"control_vs_faexcess\"\n",
    "    group1 <- tolower(split_name[1])\n",
    "    group2 <- \"faexcess\"\n",
    "  }\n",
    "\n",
    "  # Read in data\n",
    "  df <- read.csv(file)\n",
    "    \n",
    "  # Filter to only significant DEGs\n",
    "  df <- df[df$adj.P.Val < 0.05, ]\n",
    "    \n",
    "  # Filter for only downregulated genes\n",
    "  tmp <- df[df$logFC < 0, ]\n",
    "  \n",
    "  # Check if there are any downregulated genes\n",
    "  if (nrow(tmp) == 0) {\n",
    "    cat(\"No significant downregulated genes found for:\", group1, \"vs\", group2, \"\\n\")\n",
    "    next\n",
    "  }\n",
    "  \n",
    "  # Perform the enrichR analysis on the gene list\n",
    "  enrichr_results <- enrichr(tmp$external_gene_name, c(\"GO_Biological_Process_2023\",\n",
    "                                                       \"GO_Cellular_Component_2023\",\n",
    "                                                       \"GO_Molecular_Function_2023\",\n",
    "                                                       \"KEGG_2019_Mouse\",\n",
    "                                                       \"Panther_2016\",\n",
    "                                                       \"Reactome_2016\",\n",
    "                                                       \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\"))\n",
    "\n",
    "  # Log the number of enrichment results\n",
    "  for (i in seq_along(enrichr_results)) {\n",
    "    if (!is.null(enrichr_results[[i]])) {\n",
    "      cat(\"Enrichment results for:\", names(enrichr_results)[i], \"- Rows:\", nrow(enrichr_results[[i]]), \"\\n\")\n",
    "    } else {\n",
    "      cat(\"No results for:\", names(enrichr_results)[i], \"\\n\")\n",
    "    }\n",
    "  }\n",
    "\n",
    "  # Define the output file paths for each enrichment plot\n",
    "  output_files <- list(\n",
    "    GO_Biological_Process_2023 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_GO_Biological_Process_2023_down.pdf\"),\n",
    "    GO_Cellular_Component_2023 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_GO_Cellular_Component_2023_down.pdf\"),\n",
    "    GO_Molecular_Function_2023 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_GO_Molecular_Function_2023_down.pdf\"),\n",
    "    KEGG_2021_Human = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_KEGG_2019_Mouse_down.pdf\"),\n",
    "    Panther_2016 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_Panther_2016_down.pdf\"),\n",
    "    Reactome_2016 = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_Reactome_2016_down.pdf\"),\n",
    "    RNA_Seq_Disease_Gene_and_Drug_Signatures_from_GEO = paste0(\"05_DEGs/\", group1, \"_vs_\", group2, \"_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_down.pdf\")\n",
    "  )\n",
    "\n",
    "  # Loop through each output file and plot the results\n",
    "  for (i in seq_along(enrichr_results)) {\n",
    "    if (!is.null(enrichr_results[[i]]) && nrow(enrichr_results[[i]]) > 0) {\n",
    "      pdf(output_files[[i]], height = 7, width = 15)\n",
    "      p <- plotEnrich(enrichr_results[[i]], showTerms = 25, numChar = 75, y = \"Count\", orderBy = \"P.value\") +\n",
    "        ggtitle(paste(names(output_files)[i], \"for DEGs Downregulated in\", group1, \"vs\", group2))\n",
    "      print(p)\n",
    "      dev.off()\n",
    "      \n",
    "      # Print message indicating the plot was saved\n",
    "      cat(\"Plot saved for:\", names(output_files)[i], \"for\", group1, \"vs\", group2, \"->\", output_files[[i]], \"\\n\")\n",
    "    } else {\n",
    "      cat(\"No results to plot for:\", names(output_files)[i], \"for\", group1, \"vs\", group2, \"\\n\")\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
