{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "77e43832-fec1-48b2-8107-3c01a102238b",
   "metadata": {},
   "source": [
    "# Notes for Heatmap Adjustments\n",
    "\n",
    "## Legend Position\n",
    "\n",
    "**First Number: X-Axis Position**\n",
    "- **0** = Far left edge of plotting area\n",
    "- **1** = Far right edge of plotting area  \n",
    "- **Values < 0** = Left outside the plotting area\n",
    "- **Values > 1** = Right outside the plotting area\n",
    "- **1.08** = **8% to the RIGHT** of the plotting area\n",
    "\n",
    "**Second Number (0.915): Y-Axis Position**\n",
    "- **0** = Bottom edge of plotting area\n",
    "- **1** = Top edge of plotting area\n",
    "- **Values < 0** = Below the plotting area\n",
    "- **Values > 1** = Above the plotting area  \n",
    "- **0.915** = **91.5% up** from bottom (near the top)\n",
    "\n",
    "## Color Bar Position\n",
    "\n",
    "**First Number: Top Margin**\n",
    "- **Positive values**: Add space above → color bar moves **UP** away from heatmap\n",
    "- **Negative values**: Reduce space → color bar moves **DOWN** closer to heatmap\n",
    "- **Increasing**: Moves color bar **further UP** from heatmap  \n",
    "- **Decreasing**: Moves color bar **closer DOWN** to heatmap\n",
    "\n",
    "**Second Number: Right Margin**\n",
    "- **Positive values**: Add space to the right → color bar extends **RIGHT**\n",
    "- **Negative values**: Reduce right space → color bar shrinks **LEFT**\n",
    "- **Increasing**: Color bar extends **further RIGHT**\n",
    "- **Decreasing**: Color bar retracts **LEFT**\n",
    "\n",
    "**Third Number: Bottom Margin**\n",
    "- **Positive values**: Add space below → trait labels move **DOWN** away from color bar\n",
    "- **Negative values**: Reduce space below → trait labels move **UP** closer to color bar\n",
    "- **Increasing**: Creates **more space** between color bar and trait labels\n",
    "- **Decreasing**: Brings trait labels **closer** to color bar\n",
    "\n",
    "**Fourth Number: Left Margin**\n",
    "- **Positive values**: Add space to the left → color bar moves **RIGHT**\n",
    "- **Negative values**: Reduce left space → color bar moves **LEFT**\n",
    "- **Increasing**: Color bar shifts **RIGHT**\n",
    "- **Decreasing**: Color bar shifts **LEFT**\n",
    "\n",
    "\n",
    "**Movement Summary:**\n",
    "\n",
    "```\n",
    "To move color bar:    Adjust:    Direction:\n",
    "UP                   1st ↑      Away from heatmap\n",
    "DOWN                 1st ↓      Closer to heatmap\n",
    "RIGHT                2nd ↑      Extends right\n",
    "LEFT                 2nd ↓      Retracts left\n",
    "LABELS DOWN          3rd ↑      Labels move down\n",
    "LABELS UP            3rd ↓      Labels move up  \n",
    "BAR RIGHT            4th ↑      Bar shifts right\n",
    "BAR LEFT             4th ↓      Bar shifts left\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "421e76ba-d77e-4fa6-9a3a-90d352eb7a08",
   "metadata": {},
   "source": [
    "# Set Library Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ce2b8b5f-ea46-418a-8648-5a7d5262bdd9",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/home/vhaghani/anaconda3/envs/jupyter_nb_R4.4.2/lib/R/library\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32c5e124-7598-4058-b58e-e1dbc909cc75",
   "metadata": {},
   "source": [
    "# Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "35babafb-8f17-41e9-885c-1a0a74977267",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(enrichR)\n",
    "library(dplyr)\n",
    "library(rtracklayer)\n",
    "library(reshape2)\n",
    "library(openxlsx)\n",
    "library(readxl)\n",
    "library(glue)\n",
    "library(ggplot2)\n",
    "library(viridis)\n",
    "library(tidyr)\n",
    "library(comethyl)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8f3adef-dcfe-405a-994b-478a732aa1ee",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3bf35391-85c2-489e-aba7-8fe48945d544",
   "metadata": {},
   "outputs": [],
   "source": [
    "modules <- readRDS(\"Modules.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d38a0648-fdc2-4461-b3f3-967c999a914d",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions <- modules$regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9370c9f2-0b01-466c-aa23-b25a87098166",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Module eigengenes\n",
    "MEs <- modules$MEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "805fa139-0723-4f31-afe0-5dcb51538bc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample trait data\n",
    "colData <- openxlsx::read.xlsx(\"sample_info.xlsx\", rowNames = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd75abf5-5093-4e7c-a54f-bd81878ec779",
   "metadata": {},
   "source": [
    "# Make Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e23ac921-4f83-4b78-ba05-18279be11db9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[getMEtraitCor] Testing associations between module eigennodes and sample traits using bicor correlation\n",
      "\n",
      "[getMEtraitCor] Saving file as ME_Trait_Correlation_Stats.txt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Test correlations between module eigennodes and sample traits\n",
    "MEtraitCor <- getMEtraitCor(MEs, colData = colData, corType = \"bicor\",\n",
    "                            file = \"ME_Trait_Correlation_Stats.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "818fcd10-b017-4e4d-b98e-5f292b88f67c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[getDendro] Clustering with bicor correlation as the distance\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Module dendrogram\n",
    "moduleDendro <- getDendro(MEs, distance = \"bicor\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5995a2b2-5f35-4536-a55b-61cc8b5d2b82",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[getDendro] Transposing data\n",
      "\n",
      "[getCor] Calculating correlations using bicor correlation\n",
      "\n",
      "[getDendro] Clustering by euclidean distance\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Trait dendrogram with error handling\n",
    "traitDendro <- tryCatch({\n",
    "  getCor(MEs, y = colData, corType = \"bicor\", robustY = FALSE) %>%\n",
    "    getDendro(transpose = TRUE)\n",
    "}, error = function(e) {\n",
    "  cat(\"Bicor failed for trait dendrogram, trying pearson correlation...\\n\")\n",
    "  getCor(MEs, y = colData, corType = \"pearson\") %>%\n",
    "    getDendro(transpose = TRUE)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a91b299d-f255-452c-a722-f08e99ddcedf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter MEtraitCor for significant modules only (p-values less than 0.05)\n",
    "modules_of_interest <- MEtraitCor %>%\n",
    "    filter(p < 0.05) %>%\n",
    "    distinct(module) \n",
    "\n",
    "# Create module list\n",
    "modules_of_interest <- as.character(unique(modules_of_interest$module))\n",
    "\n",
    "# Remove grey if present\n",
    "modules_of_interest <- modules_of_interest[modules_of_interest != \"grey\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4ac640ad-5d93-4210-8c98-730d50c254de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"1 yellowgreen\"\n",
      "[1] \"2 lightsteelblue1\"\n",
      "[1] \"3 salmon4\"\n",
      "[1] \"4 magenta\"\n",
      "[1] \"5 white\"\n",
      "[1] \"6 purple\"\n",
      "[1] \"7 darkmagenta\"\n",
      "[1] \"8 lightcyan1\"\n",
      "[1] \"9 yellow\"\n",
      "[1] \"10 lightcyan\"\n",
      "[1] \"11 honeydew1\"\n",
      "[1] \"12 lightpink4\"\n",
      "[1] \"13 black\"\n",
      "[1] \"14 brown\"\n",
      "[1] \"15 brown4\"\n",
      "[1] \"16 lightgreen\"\n"
     ]
    }
   ],
   "source": [
    "# Confirm subsetting\n",
    "for (i in seq_along(modules_of_interest)) {\n",
    "    module <- modules_of_interest[i]\n",
    "    print(paste(i, module))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "aa4462a4-bab3-4a5f-9059-9295a6f9945f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>module</th><th scope=col>trait</th><th scope=col>nObs</th><th scope=col>bicor</th><th scope=col>Z</th><th scope=col>t</th><th scope=col>p</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ivory</td><td>Dam1  </td><td>12</td><td> 0.28682864</td><td> 0.9332108</td><td>0.9468152</td><td>0.3660501</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ivory</td><td>Dam2  </td><td>12</td><td>-0.52958538</td><td>-1.8643801</td><td>1.9742799</td><td>0.0765977</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ivory</td><td>Dam6  </td><td>12</td><td> 0.03324189</td><td> 0.1051588</td><td>0.1051782</td><td>0.9183142</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ivory</td><td>Dam7  </td><td>12</td><td> 0.18871584</td><td> 0.6040116</td><td>0.6076910</td><td>0.5569400</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ivory</td><td>Male  </td><td>12</td><td>-0.35441619</td><td>-1.1715776</td><td>1.1985638</td><td>0.2583300</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ivory</td><td>Female</td><td>12</td><td> 0.35441619</td><td> 1.1715776</td><td>1.1985638</td><td>0.2583300</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & module & trait & nObs & bicor & Z & t & p\\\\\n",
       "  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ivory & Dam1   & 12 &  0.28682864 &  0.9332108 & 0.9468152 & 0.3660501\\\\\n",
       "\t2 & ivory & Dam2   & 12 & -0.52958538 & -1.8643801 & 1.9742799 & 0.0765977\\\\\n",
       "\t3 & ivory & Dam6   & 12 &  0.03324189 &  0.1051588 & 0.1051782 & 0.9183142\\\\\n",
       "\t4 & ivory & Dam7   & 12 &  0.18871584 &  0.6040116 & 0.6076910 & 0.5569400\\\\\n",
       "\t5 & ivory & Male   & 12 & -0.35441619 & -1.1715776 & 1.1985638 & 0.2583300\\\\\n",
       "\t6 & ivory & Female & 12 &  0.35441619 &  1.1715776 & 1.1985638 & 0.2583300\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | module &lt;fct&gt; | trait &lt;fct&gt; | nObs &lt;dbl&gt; | bicor &lt;dbl&gt; | Z &lt;dbl&gt; | t &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | ivory | Dam1   | 12 |  0.28682864 |  0.9332108 | 0.9468152 | 0.3660501 |\n",
       "| 2 | ivory | Dam2   | 12 | -0.52958538 | -1.8643801 | 1.9742799 | 0.0765977 |\n",
       "| 3 | ivory | Dam6   | 12 |  0.03324189 |  0.1051588 | 0.1051782 | 0.9183142 |\n",
       "| 4 | ivory | Dam7   | 12 |  0.18871584 |  0.6040116 | 0.6076910 | 0.5569400 |\n",
       "| 5 | ivory | Male   | 12 | -0.35441619 | -1.1715776 | 1.1985638 | 0.2583300 |\n",
       "| 6 | ivory | Female | 12 |  0.35441619 |  1.1715776 | 1.1985638 | 0.2583300 |\n",
       "\n"
      ],
      "text/plain": [
       "  module trait  nObs bicor       Z          t         p        \n",
       "1 ivory  Dam1   12    0.28682864  0.9332108 0.9468152 0.3660501\n",
       "2 ivory  Dam2   12   -0.52958538 -1.8643801 1.9742799 0.0765977\n",
       "3 ivory  Dam6   12    0.03324189  0.1051588 0.1051782 0.9183142\n",
       "4 ivory  Dam7   12    0.18871584  0.6040116 0.6076910 0.5569400\n",
       "5 ivory  Male   12   -0.35441619 -1.1715776 1.1985638 0.2583300\n",
       "6 ivory  Female 12    0.35441619  1.1715776 1.1985638 0.2583300"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] ivory           sienna3         yellowgreen     lightsteelblue1\n",
      " [5] salmon4         magenta         paleturquoise   cyan           \n",
      " [9] darkolivegreen  darkgrey        green           orangered4     \n",
      "[13] grey60          white           salmon          darkturquoise  \n",
      "[17] purple          blue            darkmagenta     skyblue        \n",
      "[21] lightcyan1      lightyellow     yellow          pink           \n",
      "[25] royalblue       maroon          plum1           lightcyan      \n",
      "[29] red             thistle2        bisque4         skyblue3       \n",
      "[33] darkslateblue   navajowhite2    honeydew1       plum2          \n",
      "[37] lavenderblush3  lightpink4      turquoise       saddlebrown    \n",
      "[41] thistle1        mediumpurple3   darkorange      midnightblue   \n",
      "[45] darkgreen       black           brown           brown4         \n",
      "[49] palevioletred3  floralwhite     darkorange2     steelblue      \n",
      "[53] lightgreen      orange          greenyellow     tan            \n",
      "[57] darkred         violet         \n",
      "58 Levels: ivory sienna3 yellowgreen lightsteelblue1 salmon4 ... violet\n"
     ]
    }
   ],
   "source": [
    "# View MEtraitCor\n",
    "head(MEtraitCor)\n",
    "\n",
    "# See module list\n",
    "print(unique(MEtraitCor$module))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3d8592f8-6161-4f97-99ec-3bcad66516d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out modules that are not of interest\n",
    "MEtraitCor <- MEtraitCor %>%\n",
    "  mutate(module = as.character(module)) %>%\n",
    "  filter(module %in% modules_of_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "334c46ed-2a6e-47d6-b314-c9936d1c9ee4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>module</th><th scope=col>trait</th><th scope=col>nObs</th><th scope=col>bicor</th><th scope=col>Z</th><th scope=col>t</th><th scope=col>p</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>yellowgreen</td><td>Dam1  </td><td>12</td><td>-0.74820065</td><td>-3.06378822</td><td>3.56611335</td><td>0.005128801</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>yellowgreen</td><td>Dam2  </td><td>12</td><td> 0.55663684</td><td> 1.98574229</td><td>2.11884162</td><td>0.060140746</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>yellowgreen</td><td>Dam6  </td><td>12</td><td> 0.25789306</td><td> 0.83436706</td><td>0.84408181</td><td>0.418358325</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>yellowgreen</td><td>Dam7  </td><td>12</td><td> 0.03594185</td><td> 0.11370710</td><td>0.11373161</td><td>0.911701515</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>yellowgreen</td><td>Male  </td><td>12</td><td>-0.01151168</td><td>-0.03640473</td><td>0.03640554</td><td>0.971675481</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>yellowgreen</td><td>Female</td><td>12</td><td> 0.01151168</td><td> 0.03640473</td><td>0.03640554</td><td>0.971675481</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & module & trait & nObs & bicor & Z & t & p\\\\\n",
       "  & <chr> & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & yellowgreen & Dam1   & 12 & -0.74820065 & -3.06378822 & 3.56611335 & 0.005128801\\\\\n",
       "\t2 & yellowgreen & Dam2   & 12 &  0.55663684 &  1.98574229 & 2.11884162 & 0.060140746\\\\\n",
       "\t3 & yellowgreen & Dam6   & 12 &  0.25789306 &  0.83436706 & 0.84408181 & 0.418358325\\\\\n",
       "\t4 & yellowgreen & Dam7   & 12 &  0.03594185 &  0.11370710 & 0.11373161 & 0.911701515\\\\\n",
       "\t5 & yellowgreen & Male   & 12 & -0.01151168 & -0.03640473 & 0.03640554 & 0.971675481\\\\\n",
       "\t6 & yellowgreen & Female & 12 &  0.01151168 &  0.03640473 & 0.03640554 & 0.971675481\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | module &lt;chr&gt; | trait &lt;fct&gt; | nObs &lt;dbl&gt; | bicor &lt;dbl&gt; | Z &lt;dbl&gt; | t &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | yellowgreen | Dam1   | 12 | -0.74820065 | -3.06378822 | 3.56611335 | 0.005128801 |\n",
       "| 2 | yellowgreen | Dam2   | 12 |  0.55663684 |  1.98574229 | 2.11884162 | 0.060140746 |\n",
       "| 3 | yellowgreen | Dam6   | 12 |  0.25789306 |  0.83436706 | 0.84408181 | 0.418358325 |\n",
       "| 4 | yellowgreen | Dam7   | 12 |  0.03594185 |  0.11370710 | 0.11373161 | 0.911701515 |\n",
       "| 5 | yellowgreen | Male   | 12 | -0.01151168 | -0.03640473 | 0.03640554 | 0.971675481 |\n",
       "| 6 | yellowgreen | Female | 12 |  0.01151168 |  0.03640473 | 0.03640554 | 0.971675481 |\n",
       "\n"
      ],
      "text/plain": [
       "  module      trait  nObs bicor       Z           t          p          \n",
       "1 yellowgreen Dam1   12   -0.74820065 -3.06378822 3.56611335 0.005128801\n",
       "2 yellowgreen Dam2   12    0.55663684  1.98574229 2.11884162 0.060140746\n",
       "3 yellowgreen Dam6   12    0.25789306  0.83436706 0.84408181 0.418358325\n",
       "4 yellowgreen Dam7   12    0.03594185  0.11370710 0.11373161 0.911701515\n",
       "5 yellowgreen Male   12   -0.01151168 -0.03640473 0.03640554 0.971675481\n",
       "6 yellowgreen Female 12    0.01151168  0.03640473 0.03640554 0.971675481"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"yellowgreen\"     \"lightsteelblue1\" \"salmon4\"         \"magenta\"        \n",
      " [5] \"white\"           \"purple\"          \"darkmagenta\"     \"lightcyan1\"     \n",
      " [9] \"yellow\"          \"lightcyan\"       \"honeydew1\"       \"lightpink4\"     \n",
      "[13] \"black\"           \"brown\"           \"brown4\"          \"lightgreen\"     \n"
     ]
    }
   ],
   "source": [
    "# View MEtraitCor\n",
    "head(MEtraitCor)\n",
    "\n",
    "# See module list\n",
    "print(unique(MEtraitCor$module))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8a0313cb-d965-4bc1-8ed4-892cd8ec4655",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define traits to remove\n",
    "traits_to_remove <- c(\"Dam1\", \"Dam2\", \"Dam6\", \"Dam7\", \"Female\", \"Male\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "72781310-1719-4fd3-88d3-e9ba0aca3732",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove modules that are significant only in traits to remove\n",
    "modules_to_keep <- MEtraitCor %>%\n",
    "  filter(!trait %in% traits_to_remove) %>%\n",
    "  group_by(module) %>%\n",
    "  filter(any(p < 0.05)) %>%\n",
    "  pull(module) %>%\n",
    "  unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0546008b-abf7-4073-9aac-b51cf16d6c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter MEtraitCor to keep only these modules\n",
    "MEtraitCor <- MEtraitCor %>%\n",
    "  filter(module %in% modules_to_keep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3694ef96-87f6-4c26-bbb1-77682812b7b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove all traits to remove from MEtraitCor\n",
    "MEtraitCor <- MEtraitCor %>%\n",
    "  mutate(trait = as.character(trait)) %>%\n",
    "  filter(!trait %in% traits_to_remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8d8a4325-50b0-4a9d-bc03-7544819176a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Remaining modules after filtering:\"\n",
      "[1] \"white\"     \"purple\"    \"honeydew1\" \"brown\"     \"brown4\"   \n",
      "[1] \"Remaining traits after filtering:\"\n",
      "[1] \"Control\"   \"FA_Excess\"\n"
     ]
    }
   ],
   "source": [
    "print(\"Remaining modules after filtering:\")\n",
    "remaining_modules <- unique(MEtraitCor$module)\n",
    "print(remaining_modules)\n",
    "\n",
    "print(\"Remaining traits after filtering:\")\n",
    "remaining_traits <- unique(MEtraitCor$trait)\n",
    "print(remaining_traits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "edee362b-b42f-4294-b865-a1992e47355a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"white\"     \"purple\"    \"honeydew1\" \"brown\"     \"brown4\"   \n",
      "[1] \"Control\"   \"FA_Excess\"\n"
     ]
    }
   ],
   "source": [
    "# Set axis variable labels\n",
    "moduleOrder <- unique(MEtraitCor$module)\n",
    "print(moduleOrder)\n",
    "\n",
    "sorted_trait_order <- unique(MEtraitCor$trait)\n",
    "print(sorted_trait_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6b0a65ab-0de3-46aa-90d9-c1893e91d655",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to factors with explicit levels\n",
    "MEtraitCor$module <- factor(MEtraitCor$module, levels = unique(MEtraitCor$module))\n",
    "MEtraitCor$trait <- factor(MEtraitCor$trait, levels = unique(MEtraitCor$trait))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "81ed6f74-bff7-4a93-b3f4-85f8293f11b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>module</th><th scope=col>trait</th><th scope=col>nObs</th><th scope=col>bicor</th><th scope=col>Z</th><th scope=col>t</th><th scope=col>p</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>white    </td><td>Control  </td><td>12</td><td>-0.5851837</td><td>-2.119706</td><td>2.282047</td><td>0.04562667</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>white    </td><td>FA_Excess</td><td>12</td><td> 0.5851837</td><td> 2.119706</td><td>2.282047</td><td>0.04562667</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>purple   </td><td>Control  </td><td>12</td><td> 0.5762287</td><td> 2.076978</td><td>2.229562</td><td>0.04987953</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>purple   </td><td>FA_Excess</td><td>12</td><td>-0.5762287</td><td>-2.076978</td><td>2.229562</td><td>0.04987953</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>honeydew1</td><td>Control  </td><td>12</td><td>-0.6325559</td><td>-2.358001</td><td>2.582672</td><td>0.02729093</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>honeydew1</td><td>FA_Excess</td><td>12</td><td> 0.6325559</td><td> 2.358001</td><td>2.582672</td><td>0.02729093</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>brown    </td><td>Control  </td><td>12</td><td> 0.6128433</td><td> 2.256166</td><td>2.452505</td><td>0.03411121</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>brown    </td><td>FA_Excess</td><td>12</td><td>-0.6128433</td><td>-2.256166</td><td>2.452505</td><td>0.03411121</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>brown4   </td><td>Control  </td><td>12</td><td> 0.5795960</td><td> 2.092967</td><td>2.249153</td><td>0.04824897</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>brown4   </td><td>FA_Excess</td><td>12</td><td>-0.5795960</td><td>-2.092967</td><td>2.249153</td><td>0.04824897</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & module & trait & nObs & bicor & Z & t & p\\\\\n",
       "  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & white     & Control   & 12 & -0.5851837 & -2.119706 & 2.282047 & 0.04562667\\\\\n",
       "\t2 & white     & FA\\_Excess & 12 &  0.5851837 &  2.119706 & 2.282047 & 0.04562667\\\\\n",
       "\t3 & purple    & Control   & 12 &  0.5762287 &  2.076978 & 2.229562 & 0.04987953\\\\\n",
       "\t4 & purple    & FA\\_Excess & 12 & -0.5762287 & -2.076978 & 2.229562 & 0.04987953\\\\\n",
       "\t5 & honeydew1 & Control   & 12 & -0.6325559 & -2.358001 & 2.582672 & 0.02729093\\\\\n",
       "\t6 & honeydew1 & FA\\_Excess & 12 &  0.6325559 &  2.358001 & 2.582672 & 0.02729093\\\\\n",
       "\t7 & brown     & Control   & 12 &  0.6128433 &  2.256166 & 2.452505 & 0.03411121\\\\\n",
       "\t8 & brown     & FA\\_Excess & 12 & -0.6128433 & -2.256166 & 2.452505 & 0.03411121\\\\\n",
       "\t9 & brown4    & Control   & 12 &  0.5795960 &  2.092967 & 2.249153 & 0.04824897\\\\\n",
       "\t10 & brown4    & FA\\_Excess & 12 & -0.5795960 & -2.092967 & 2.249153 & 0.04824897\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 7\n",
       "\n",
       "| <!--/--> | module &lt;fct&gt; | trait &lt;fct&gt; | nObs &lt;dbl&gt; | bicor &lt;dbl&gt; | Z &lt;dbl&gt; | t &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | white     | Control   | 12 | -0.5851837 | -2.119706 | 2.282047 | 0.04562667 |\n",
       "| 2 | white     | FA_Excess | 12 |  0.5851837 |  2.119706 | 2.282047 | 0.04562667 |\n",
       "| 3 | purple    | Control   | 12 |  0.5762287 |  2.076978 | 2.229562 | 0.04987953 |\n",
       "| 4 | purple    | FA_Excess | 12 | -0.5762287 | -2.076978 | 2.229562 | 0.04987953 |\n",
       "| 5 | honeydew1 | Control   | 12 | -0.6325559 | -2.358001 | 2.582672 | 0.02729093 |\n",
       "| 6 | honeydew1 | FA_Excess | 12 |  0.6325559 |  2.358001 | 2.582672 | 0.02729093 |\n",
       "| 7 | brown     | Control   | 12 |  0.6128433 |  2.256166 | 2.452505 | 0.03411121 |\n",
       "| 8 | brown     | FA_Excess | 12 | -0.6128433 | -2.256166 | 2.452505 | 0.03411121 |\n",
       "| 9 | brown4    | Control   | 12 |  0.5795960 |  2.092967 | 2.249153 | 0.04824897 |\n",
       "| 10 | brown4    | FA_Excess | 12 | -0.5795960 | -2.092967 | 2.249153 | 0.04824897 |\n",
       "\n"
      ],
      "text/plain": [
       "   module    trait     nObs bicor      Z         t        p         \n",
       "1  white     Control   12   -0.5851837 -2.119706 2.282047 0.04562667\n",
       "2  white     FA_Excess 12    0.5851837  2.119706 2.282047 0.04562667\n",
       "3  purple    Control   12    0.5762287  2.076978 2.229562 0.04987953\n",
       "4  purple    FA_Excess 12   -0.5762287 -2.076978 2.229562 0.04987953\n",
       "5  honeydew1 Control   12   -0.6325559 -2.358001 2.582672 0.02729093\n",
       "6  honeydew1 FA_Excess 12    0.6325559  2.358001 2.582672 0.02729093\n",
       "7  brown     Control   12    0.6128433  2.256166 2.452505 0.03411121\n",
       "8  brown     FA_Excess 12   -0.6128433 -2.256166 2.452505 0.03411121\n",
       "9  brown4    Control   12    0.5795960  2.092967 2.249153 0.04824897\n",
       "10 brown4    FA_Excess 12   -0.5795960 -2.092967 2.249153 0.04824897"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(MEtraitCor, n = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "f26b084e-53f2-47d0-bef0-70f5eac4fae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[plotMEtraitCor] Plotting ME trait correlation heatmap\n",
      "\n",
      "[plotMEtraitCor] Using colors in column names for x-axis labels\n",
      "\n",
      "[plotMEtraitCor] Saving plot as ME_Trait_Correlation_Heatmap.pdf\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdZ3wVVcLA4XPTEzqIEOkIFlAE7BVUxN6wl9W16762tWJvq4JtV1d3LVh3FVl7\nb6jYsKAgYgMVEOkiJZT03PcDqIAmAklM7vF5fvlAZiYz58Yx+WfKnUQymQwAAKS+tLoeAAAA\nNUPYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgB\nAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY\nAQBEIqOuB8Aq2fTQC0cPu7auRwHAH1EymazrIbCqHLEDAIiEI3YpJqfb7lld+tT1KFLbhZ9d\nWddDiMdlG1xa10OIxxXjLqrrIUQkI6uuR5Dy3phf9sLc0roeBatN2KWYrC59GvU7v65HkdrO\nX3hdXQ8hHlfvaG+sMef/cEldDyEiOTl1PYIIFAm7VORULABAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSE3e/h44s3SCQS+/23qK4HAgDErL6HXeG3w287/6jdNuuS36JhdnaDtdp22Wjz\nvgP+75r/vD5hQXldDw4AoD6pz2G34P0b91mv6y6nXveflz76ZubcxSUlS36Y9s1nH77xxL8u\nOmqn3me/UvObfO+cjolE4sCHy2p+1QAAtSyjrgdQmdJPBu+588B3Fofcdfc89by/HNxv8/Xb\nNM1cMmfGjBlTvxr10uP/e6FpvR07AEBdqK9xNP7Goy96Z3FouP3fRjx/0aYNf5ycnd+pWX6n\nbr233/ekv9Xl8AAA6p/6eSq24tV//P3j8pC1/dUPLld1lSud+votp+6zxbotG2Vn5TZt063v\n4RcPHbdg+SWGn7xWIpHz52eTc0bednL/jddpkpPdoGXXbY+4+uWpP16qN/z4pomtb/w2hPDY\nYZmJZVqf+fbyX14+47W/n9C/R9vmeRmJzW6YvKpbBwD4HdTPI3afvPLK7BDS+x/353a/vXDp\nJzfvsfNfh89JLvu8ZPoXbwy9+o3H/zfikbfu2LvV8ot+/9yJW9w1ZNKylCv+euRDF+/x3uzh\nn9zct8GqDGvWE3/e/N7/Tlu2oYqKitXbetVGjhz51ltvVTZ38Q/TVn1VAMAfU70Mu4qvvpoU\nQujcq1fj31w2+emgI84ePifZaJPjBv/97H02bZ8754vX773g9GuHf3Xnn07ZedLjBzf7adni\n529/eOM//eOl8/bfonOLku/euefMYy54ceK/r33osr4nNA/9hsxPHn9Ox61v/PaAoaWPHvrL\n70zxi/c8tt5B1z130UFbbdCheXZiNbf+G1599dVLL720srnr9Oi7qisCAP6o6uWp2IUFBRUh\nhMaNf7vrwjtD7vi0PDTb//bhQ07ZccM2jRs077zZAVc9/8w5G6aFBU/ePnTm8gs32eOWlx84\no/9G7ZvmNVh7/f4D/3v5Tmmh9L33Plq1cWVvfdXzw87dY5OOS6tudbcOAFCr6mXYNWrcOC2E\nUFBQ8JuLzvzww2kh5O13/GFrLT85s/cpx20RQnLUqBWabdsDD2i9/OctevRYJ4SCOXNKV2lc\nWx52+LqJNd86AECtqpenYtO6du0UwjcTx4wpCD2qPmq3YMGCEELbdu0SK81o265dIry3aP78\n8hDSl03LbtlypbVlZ2eHEMrLy0PI/M1hZXbokF+Nrf+GAw44YIMNNqhs7uX3Dp/+yYhVWxMA\n8AdVL8Mu9Nhll7UHfzP75Xvun3b0aW2qWrJJkyYhhKnffZcMG69QV1O/+y4ZQsOmTVexq1ZB\nWtrKxzdrcuvdunXr1q1bZXMHPTZmtcYKAPwB1ctTsSFtpzPO2CQtlLx50ZGDxyyuasnWm23W\nJoQlTw4ZOmf5yWUf33HvqBASm2226WptODMzM4RQVLSKD3Wt4a0DAFRH/Qy7kNjgnPuu2io3\nLBwxcLve+1ww5MXRk+csKikrmj9z8hdj3n76zkv+3G/Tc4eHEMK2x53YPS3Me+LkXU64440v\nZyxcPHfS6Ccu23Pv6z8rD433Pfmw1r+1qRW0bt06hPDB0w99MadwVR4rVrNbBwCojvp5KjaE\nkNXzwueHpx9x0MUvTHhm0AnPDPrFAg2OOyeEEBIbD/zv9SN2Pvv1j4ec3HfIycutYN1j7//X\nIc1Xb6tt9tin9zlvj37ipG5PnBRCCKHVGW/N/Md2lS5fs1sHAKiOenrELoQQQrNtzn/+qwkv\n/fOcI3bp1blV07zMzLzm63TuvmmfA/7vmv+8NubGXZYultXzrJfGDP/7KXtu2rFFg8yM7Mat\nN9jh0Av+8/6ou/fLr3oDv6LrmY88cuH+W667VoPMVfvW1OjWAQCqIZFMJn97KerapodeOHrY\ntSGExvsMatTv/LoeTmqb+kSLuh5CPBrv+UNdDyEeBY9m1fUQIpKzCk+jpEqDpxQNnFi49N9S\nIYXU5yN2AACsBmEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcA\nEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEH\nABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlh\nBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJ\nYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcA\nEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwDEq+LThy+//PJbhs+s64H8PjLq\negAAAKtr4tODHhhd9PPniYzshs3brNe7z85bdshLLLdgxacPX3HFU+uev9/p/Vr/7qP8/Qk7\nACDlTHx60BV3L/jl9ETTnsfcNuzfh6+XtWxC2kaHXnZZz+bb/RGqLgg7ACBl5e9yxonbNA0h\nhIri+bMnjXnxmTe//fieP+2z9gafX9t76eVmaRsdevlGdTrI35WwAwBS1Dr9z7z8nI4/f178\n+bXb97xw1Phhj358be/edTasOiTsAIBIZHfbs1/HC0d9tWTJkh8nVXz68JWPftl8u5NXvsau\nZPanb73+7qdT5hTntO684eZ9dtyoZfoKCxRNH/3aq+9+MXVBWW7LTj377LL9es2WX2Dis9c9\n8GFJz8Mv3q/rkinvvvjyh1/PnNd0px+PINYVYQcARKLkyxdenRxCTu/eG/446dduniiZ9MTA\nY0677Y1pJT9/aW6nftc8/sqZPZd+0YyXLjr82BtHTC/9eYGGGx42+OG7/tKjwbLPJz573RV3\nLDq6y47jTj7w8tdnVoQQQu/so4QdAMCamP7yPy5ftOwau4Lvv/345WdGTCxv1f/GQYe1qPyL\nZj5+1PYHDpuWDDkdttl9h03Wzc+aO+nzUSNeG/721+HMniGE4g8v3nWfQeNKQm6nHfbeedP2\nuXO+ePP5F8Z+MfT/dl7UcOzTR63z87rKXjp3v5lzMrfY7+gtN+jQPHuduq26IOwAgJQ145Wb\nr3hl+QnZnQdcd8+/TumRWemXlL525enDpiXzNj3z0ccH7d4++8fphZNeGV0RQghh1n8uvnFc\nSWi5911vDzt+vdwQQggVs149q99uN3/6zMBBbx5+yw4/5VP5zIINrh713IU9G9fwK1tTwg4A\nSFE/3xUbypb8MPObUS88+/g5O77z4nXPP3VO77xf/ZKRjzw6LYT1zrr7xt3bL/+YhtxOu2wb\nQgihcPhzr5WExPYX//vHqgshpLXaefBNx/6v/50znn129C07bPHTl3X+yw0D603VBWEHAKSs\nle+KDSVf3NBvi3OHn3/M9bt+fNnGiV9+RcE333wfQtMd+m5S2cO3pk+eXBpC2y23bLPi9Oyt\nt+4d7nxuyuTJybDFj2tObNKr0hXViXo1GACAasja8C8n9UsLFZ88/ezkX10gmUyGEBKJX2m+\nVV5i+e01bpyzBsOsPcIOAIhIMhlCCLNnz/7VuU26dm0Zwrw3Xh9bUcnXt+nUKTOEqe+9N23F\n6SXvvTcmhNC+Y8dVSr46IuwAgFiUfHHbna9WhJDWteu6v77E1gce2CaECTcdf84LU4qXm144\n+dWR34QQQm6/PXbMCsm3rj7l7q8Kf5xb8f2IC86+e3oI+XvuWa/f+Ng1dgBAivr57U5C+ZK5\nMyaOeuHZ96YXh7D2EWccttavf0nmTpf+85CnDxj24d/3WP/xbXfvs8m6rTPnTf78g9dfG7fD\nw8lt1g2h1Z/+9tebXx/82TPH9+z+37136d0ue874t55/7uM5FaHFntec36det1O9HhwAQOV+\n8XYnIaQ16fGnm/57+77NKv2i1vs/8OYjrY469d/vfPvOEw+8s2xqXud+23VZ+s+cza9+6cny\nw477x1uTRgy7c8SyBRqsf/C1Q4f8uW1Nv4aaJewAgJTTeZ+Bl7Ut+vnzRFpGdsMW7TbYrM+O\nm7fNXW7BtI0Oveyyns23W/55YlmdD7j57b0Gjnn9tfc+/25+eaN1Om+4ed8+3Vr89MSw9DZ7\nXP/mpDM/HD78vS+nLSjLXatTzz677LBB8+UfKdZ5r/Mua13Sc71ae4lrJLH03g/quU0PvXD0\nsGtDCI33GdSo3/l1PZzUNvWJKt6PnNXTeM8f6noI8Sh4NKuuhxCRnIZ1PYKUN3hK0cCJy64w\nkwopxM0TAACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACR\nEHYAAJEQdgAAkRB2AACREHYAAJHIqOsBEIvS0UuGP10WOuf2Oyozs64Hk+qWlFw5q7wiK/P0\nNhnN63osKW/24yUff5JstFPW1jsk6nosKa+w4sofkhWZaae3Stgzq2nK3KJ7CpKNG2ef1dwR\nFmqS/YkaUjZ6yQtXLHzhgZKyuh5JBJaUXDm58IrppXPreiAxmP148ctXlLz7ZrKuBxKDwuSV\nUyuumJW0Z1bflLnFV0wuumluRV0PhNgIOwCASAg7AIBICDvWRMWUF0sKSldlyWTBW0Xfza/t\n8aS0cXNLJ6/a2ZgfFpaOLK7l0aS24ooJL1es0sUAFcmpz5UX1PZ4Utu4BclV3TMXJ0eW1PJo\nUlqyYsQP5QtXbdHPfyj7xnUDVIOwY/UVD19w9+7fX9l97gtPlFXx07zkq8IXBsy+cocfhpxb\nLEcqUbKw8NBPFm7w/sLzZpUvqHyx4qLSGz5f0OXDhXtPKJn9+40u1Uy9qnDIrosH7VL80dhk\nFb8a5w8vGdp78c17FT7+oF+glSlZXH7ol2UbjC07b06yqj2zOHnD12VdxpXtPbnCnlmZr6ct\n3n1cQdfRS+4oqCivfLFZBcUnjy7oMW7RsdMq7JmsMXfFsvqKszLX2yJt1AeFLwwoemf7Bnvd\n2GjLzVf4EyH5Q8n7Vy549t8lBaUhrWXWeuuH4hCy62q49drikNa3Sdo380uv/3zBPd9lX9ol\n95SmaSvcVVxWPmzKkoHflU6uCGkZ6Qc2ShQmQ3B7569JJjukd2pX8c3wkqG9S986Kmvvv2Wt\n22aFJYo+L3vt3OI3n68oCyF3o4x1miYrQsKft79mcUj0bZT4piB5/ddl98xIu7RD2imNEyvs\nmeXJYdPLB85ITq4IaRmJA/PsmZVJlmRl7JRb/nxB8cmjS25pmXN95+w9clf4ThUWlt40sXDw\n9+ULQ2iQm7lVVrIohNy6Gm/9V7QknL1fza92/Z7h9Ouqs4Kirx6/5ur7PpiVu/4+f738lK2a\nrTS75NsXbrn+vte+nL2kbJ3D73jwxPWrs60qCDtWX+MdGh7xfoPdXlz80hULP3hr0UNbLhlx\nWKNdOydDCKG4/Msb5r109ZLp80Naq6ytzmm0619yWuTV9Yjrr2aNsm/rlX3h/OJrJxcOmVd8\nxpiSW9fKGdwkhBBCMvnOtMVHTip+vzSkZaQf2iH3krZZ3fwvW7lEuxNyTjk6e+I9xS9fW/r1\nfcX//l9p97OzOy4KIYSK2eUjTy56eUj5ovKQ2yNj50uytz8gLUeGVKpZg7TbuqVdWFBx7dSK\nIQUVZ3xecWuztMGNls5MvjOr/MipFe+XhrSMxKHrpF2Sn9YtvW7HW58luq2d+1zLnFGziq78\ntvjZ7wv3/KF45za5+5eFEEKyrPyBbwovmlY2tSI0zMs8v33uOa3S17JjVq28LIx6teZXu2qX\ncVSq9O1z+x0/9qC/n7XX3KEX7H5U3vhnjl57udnzHj9mqxMnH3jlX8/dcK3svPbrVG+sVUkk\nqzpjQX2x6aEXjh52bQih8T6DGvU7v66H87PkDy8vfvmKRR+MLF/+BEN666wtzm3c/+Tsepl0\nU59oUddD+HXTFxQPmlR417yKouUmpmWmH9I299K2WRvUy6RrvOcPdT2EX1WanHRvycvXlHz1\n7QqT8zbJ3OHSrO32r59JV/BoVl0P4ddNX1gxaGrFXQuSK+yZGYlD8tMubZ22Qf1MupyGdT2C\nX5NMfjS76Ipvi59ZssJv3kZ5mad2yD177fQW9WnHHDylaODEwqX/rl+psLgg7FwLP8Z79wn/\nGr7GX13y9FHNL13v448v7hJC4RNHtBrU89P3z23/0+xxl2+w7dgrpjxxSNOaGGqV6uXvClJH\nokX/hof1z9ts4Jw7Bpcuvd4uq1ejk4c37uLtS1fbOk2yb+mZ9ecpi/p9UzovhBBCVl72Iz0b\n7OMs9mrLTHQ6MfukIzNG7FX47OvLfiO1PCb3lCEZjZ15XW3rNEq7ZcO0P08v6zcluWzPzEl7\npFv6PvU0ROuxRGLTVrlPr51992cLT5iz7EK6Nmvljeie3aU+JV0K6Lb5Cp8uWhCmfrN6a0gk\nwvq9VpjSun0li66SaePHJ3od0iWEEEJur17rTRj/VQg/rbFg1Afjt+2Td8cxe78xr9nGe50x\n8LhNm9XWf3JhR7Uk5w5f/PLli95/5+cjdiVjFt7WvXjL8xv3Pym7uatEVsOMBcWDJxfeMffn\nI3YlS4oHjCo7tF3uJW2z1q+fx0XqqdLk5PtLXr66ZMLkn6d9f2/hDWMy+1yWtd2+adl+i66G\nGQsrBk+tuGO5I3YlRRUDPkkemp92Ses0e+ZqSCbHzC664tvip5Y7YjdtzpJNR5We1iH3rLXT\nm9sxV0ki7Hf8ChO++Sw8dvvqrSMtbeWVlFfrVOzixYtzc3/8lZeXl7d40aLl5s6bNy/j7Ttu\n73XZKafmjrvzvJ0OTX7x0gm1dDpW2LGGknNfWfzyFYvef6e8PJG2zmFNdl238N6/lYTtGh2z\nT9lLVxeO/Ouc9wdnbXle4/4ny7vfNGNB8aDJhXfOrSgKia5r5Q5qUn7wNyXlDXLua5P816Ti\nBycuevi79EPb5V7cNquenvaqR0qTk+8refmakgmTQ8hL635xdsdxhc89FVqfmrNNSenLd5e+\nsH/pGz0zd7g0a/v95N1vmrGwYtDUijsXJItC6NosbVCjcPCUivLctPtah39NrXjwu/KHZ1Qc\nmp92cb09IVt/JJOjZxdduTTp0tJ2bpu3X1nhaTMr1mmdd01GyUXTS6/+ovSWbzNP65B7trxb\nFTecufKUtNXfBVdaSa/tw/4nrvGI8vPz546dsexGwenTp7fOz19ubuv8/PL8Ha+95ugeIey1\n6ffPt3nh7dITDq6dx286K8HqW/jOoqHbzrqq/4KR71Q02L7hYe+3Ov+hhht0SIQQQnb6Buc2\nP/+bloedltXgh5KRZ825qvP3Q/9Rsmrv4PQHNH9R8Rkfz+88evEtcysaNMr+R68mn22cO2Dp\npYmJxLZtGry3VZOhHTLblZc/OHFR93cXHPFt6ZT6dK1LvZKcdm/xHV0X3XpiyYQpibZHZ58y\nocExV2W0bBhCCGmt0re5I2/g2Nyddk8r+bj0xQGLr+5VNPzFCs9zqsT8xRVnfFHW+bPyWxYk\nGzRI+0e3jM/WTx+QE0IIIZHYtlX6ez0zhq6TaFeRfPC78u5jyo6YlrRnViL55feF+45asOkX\nRU8tCd1a5j67eePhXbJ6ZoQQQiIj/egujb7avOFVLdOTS0qv+aKg46hFF31fXvRbK/1DS4SQ\nnl7zH2nVKqIWO/fr8dp/HpxWEcKSd+5/dGH//pssNzd7p936zhs3dlYyhFDw8ceTWuTn19pD\n1YUdqy9rcekXI8vTu+bs/tjal77ZZOvN01b6+zLRImvrW1pe+lnz3ffLSJ9Z8sWnSdfiVCIv\nWfHK/IpkduY5Gzb5erMGZ6z0XichJDLSD+3c6MstG13XKr1RafmLBUmHmSqTqPi67JtvQ9Od\nsw4b3eCM+1Z+r5MQQk73jD2ebzDwlezemyQKx5Z9N8d7nVQmLyRfKUgmsxLndMn4euP0M1Z6\nr5MQEumJQ9tnfLlJxnVrJRqVJV9cZM+sTCKtuOzlJclWjbP/3bvxJ91z9sxd+TuVm5t5cffG\nX/fKO7FxYsmSsjeLEy6s/Q1p6TX/Uc0fB11Ovu6EWWdv1LXnRl12f2zDGy7aNTuEj28ecNCt\n40IIofVRg88uOKdbl8236NblsLd3/9eF29fI9+HXOBXL6svu3+T4Fxo03TmrcZV/cWR1zd39\niZxt3ipesHG2n1KVyGqUO6xHRqOmmR2r/JGSnZN5brcmx7YrHZ+V2er3GlvqaXdp7vF9Quf+\naVX/YGvaL+vw0Znbv1DRZE8pUpmsBunDNkhr1CjxG3tmduLcLhnH5ifHZybsmZVZr02D53OT\nm7VIb1TlYq2aZN/RO+v0H8pzmq/8xzIrS6+Fc//p1fw7r0m/Gz6c9H9jv5id3XmT7q1yQgih\n4x7nnJVYegdFzuYDR0z806fjpiXXWb9b+ya1WF/CjjWR1n63VTwGl2iyfU6T2h1Mqtu4+aoe\nkW/RKHObWh1KystOW6//qi2Zlmi3p6vCqrZxk1WtixYNEvbMqiTSdlzVd+dIdG/h9/IqWIMr\n6n5TovrrTG/aqffWnX7+vGnXbbZebgMN2my81S9OJNQ4OxAAkEIStRJ21bvGrv4QdgBASqmV\nI3bCDpaX0Ttv98uyQucs+1T15WVd2jG9IivTuzzXgLUHZPfvkmy0g4uWakBu4tK2aRWZCXtm\n9bVvnn1ZRrKxd8xeM7VxjZ0jdrCCzN55u/eu60HEIi/r0k6/vRSrZO0BWf0H1PUgYpGbdmnb\nuh5DLNo3z7lcIK+ZRAgZtfBmIemRFFEkLwMA+KNwjV3lhB0AkFJq5VRsJDfKCzsAIIUkXGNX\nBWEHAKQUp2IrJ+wAgJTiiF3lhB0AkFJq4w5W19gBAPzeErUTdrVxFLAuCDsAIIXUziPFauBZ\nsfWCsAMAUopr7Con7ACAlOJUbOWEHQCQUtJqoV6cigUAqAMZjthVStgBACnEkyeqIuwAgNSR\nqJ1Tsd7HDgCgDjhiVzlhBwCklFp5VqwjdgAAvz83T1RO2AEAKSThWbFVEHYAQEoRdpUTdgBA\n6kjU0jV2bp4AAPj91csjdkVfPX7N1fd9MCt3/X3+evkpWzVbYWZy3kcPXH/rk2Pn5q3X95iz\nTuvXrtb6K5I+BQD+KNLTa/6jmkfsSt8+t9/xI9Y64OTjtpp50+5H3T97hbnf3HzUue+36Hfs\naUdtt/CBAXtc/Um1tlUlR+wAgBRSH2+eKHnhznubnfXxDUd3CWHX9A9bDQx7gJkAACAASURB\nVLp/ytHntv9pdpsThw3Py0sLIZS3HnPrUzO/r+5oKyfsAICUkt9+hU9LisOCH1ZzFYnQMn+F\nCU2aV2dE08aPT/Q6pEsIIYTcXr3WmzD+qxB+HmVOXl7JKxf0/9ub8ydPbXjE/57cuTrbqpqw\nAwBSSu8+K3w6Z0b4+O3VW0MisfJKykqrM6LFixfn5uYu+yQvL2/xokUrLZHR4/DLL9th1idD\n/zb48v8e2efMbrV0MZywAwBSRyKE155YeeIaPD12pZXkdwjr9VzjQeXn588dO6M4hOwQwvTp\n01vn56+0RFqrjfu22jjstFuTT5uc+8hXZ162/hpvrEpungAAUkr9u3mixc79erz2nwenVYSw\n5J37H13Yv/8my80teOvhpyaXhBBCKBw7ckxxu3YrZ1/NccQOAEghifr4SLEuJ193wsMHbdT1\nljaFEwu2vO2NXbND+PjmAVenX/HIqRvntiz5376dz0pv03TB1xPTd7z1yT81rqFR/5KwAwBS\nSq28QXE119mk3w0fTvq/sV/Mzu68SfdWOSGE0HGPc85KtA8hZG5w1IOjB0z57Is5WW3W67pO\nw9p8yIWwAwBSSnpmza+zBmIxvWmn3lt3+vnzpl232frnmQ3b99i8/a98VQ0TdgBASqnmadNf\n5VmxAAB1oDbeoLg2YrEuCDsAIHUk6uOTJ+oPYQcApJRaORUbyRvACTsAIKU4Ylc5YQcApJSM\nWrgrtjZisS5E8jIAgD+K2jgVm+5ULADA7y2xJk+G/e21OhXL72iT8c+MDiGEcOGnl5+/4No6\nHk2K63DQ/LoeQjwK7k3U9RDi0eLIZF0PIR4N/HKrtoJXBoeJA+t6FL8mEerjI8XqDft+amjd\nOJIdDgCqy80TlRN2AEBKqZVnxbrGDgDg9+dUbOWEHQCQQjx5oirCDgBIHQnX2FVF2AEAKaU2\nws6pWACAOuCIXeWEHQCQQhK1cnTNXbEAAHXAs2IrF8nLAAD+KJyKrZywAwBSRy3dFevmCQCA\nOuAau8oJOwAghXiD4qoIOwAgpQi7ygk7ACCluCu2cpGcUQYA/ijS02vho7pFVPTV45f+eZ/d\ndj/kjH+/N28159YgYQcApI5EIqRn1PxHNU/Flr59br/jR6x1wMnHbTXzpt2Pun/2asytUZEc\neAQA/ihq47RpIlGdry554c57m5318Q1Hdwlh1/QPWw26f8rR57Zftbk1S9gBACml/l0PN238\n+ESvQ7qEEELI7dVrvQnjvwqh/arNrVn17lsDAFCV6h1dqw2LFy/Ozc1d9kleXt7iRYtWeW7N\nEnYAQCopr6j5dSYSIa0auZifnz937IziELJDCNOnT2+dn7/Kc2uWmycAgFRSXlHzHxXVi8UW\nO/fr8dp/HpxWEcKSd+5/dGH//pus8tya5YgdAJAykiGUJWt+tenVTKIuJ193wsMHbdT1ljaF\nEwu2vO2NXbND+PjmAVenX/HIqRv/6txaIuwAgFRSVgunYqt9CrNJvxs+nPR/Y7+Ynd15k+6t\nckIIoeMe55yVaF/Z3Foi7ACAVFIr19jVwDrSm3bqvXWnnz9v2nWbrSufW0uEHQCQOpK1ciq2\n3t1nu6aEHQCQSmrjVGx1bomtV4QdAJAykrUTdtV+VGx9IewAgFRSG2GXURs3ZNQFYQcApJLy\nWrjGrqIW1lknhB0AkEpq44hdbcRinRB2AEAqqZWwcyoWAOB3Vks3TzhiBwBQB2rj6Fo1nxVb\nfwg7ACCVOGJXBWEHAKSO2nnyhLADAKgDtXIqVtgBAPzOksERu6oIOwAglXi7kyoIOwAgldRG\nhDliBwBQB5yKrYKwAwBSiZsnqiDsAICUUVtPnnCNHQDA7682Tps6YgcAUAfKymt+nY7YAQD8\n7pIeKVYVYQcApBJ3xVZB2AEAKSNZS3fFOhULAPD7cyq2Cml1PQAAgNVQnqyVjxoc4NSXrztx\nwG67DTjphlen/6JCF3/xyFUnHLDnPof+5dqnJhTW4GZDCMIOAEgtZRU1/1GTYffZoD0OfCjR\n/8QTdw3/PWDfv3+1wsy5D5xy/HOJbY867bj+ec8d2+e8ETV8i69TsQBAKqnn19iNvPuOecc+\ndPvJ2yXC/i0/b//nIaPOHrz5T3MbHXj7iKPyMkMIYZNJ99z6zqyiEBrU2LaFHQCQQpIhNMxe\nYUppeVhcutrraZqzwqdZ6dUa1XKKxo//rufBvRIhhJDo1WuTSU9NKA2bZ/44OzMvL3xw/e7n\nPTV7yrfJXR96+sCarLog7ACA1LJr5xU+nb4wjJiyemtI/GIlRWXVGtJz5/W9/oMQQps/Dbm5\ncHHIzV2Wjbl5eclFi5aE0GT5pbvsc+FlPWZ/+ex1V112x1F7XLV1zq+scU0JOwAglTwwruZX\n0rFJ2LDFmq+t95GXX75HCCG3Q+sWo/OzZsyYGUKbEMKM6dNz8/ObrLR08/W333H9sOOuHb7t\nsvdD71+1dZ813/AvuHkCAEgltXFLbDWfFZvfo2/fvn379t2yU8PEDv12GvPQ/eNLQygdf//Q\ncf37b7f8kh8+9sAXS0IIIZROHDlqZrt2bau14V9wxA4ASB3JUFoLz4qtwffGa3nE3y66b7ct\nugzrmJw8s8tlww9uHkL4+OYBV6df8cipG6+d/sFBG1xclN+6eNL4wp4X/+//1q2xDYcQhB0A\nkEKStfNmwtU8YreCnE0veOv7C1ac1vOMxx8JIYTQfr9b39/v1prb2MqEHQCQSmrj7U6iefKE\nsAMAUkmtPFLMs2IBAH5/tXIqtuZXWTeEHQCQSur5kyfqlrADAFJGMoSyWjhi5xo7AIA6UCvX\n2Ak7AIDfn1OxVRB2AEAqqY2ja7Xwnsd1Q9gBAKmkvr9BcZ0SdgBAKnEqtgrCDgBIGclkKKuF\n86beoBgAoA44FVsFYQcApJJaeR+7ml9l3RB2AEAqqY3Tpk7FAgDUAadiqyDsAICUkfTkiSoJ\nOwAglThiVwVhBwCkklp5HzthBwDw+6uVR4oJOwCA359TsVUQdgBAKnEqtgrCDgBIGUlH7Kok\n7ACA1OFZsVUSdgBAKnHzRBWEHQCQSmrlGrtYjtil1fUAAABWQ3my5j9qtOvKp7583YkDdttt\nwEk3vDr9l2sun/X2v846cq9dduzbd/frP6jJDQdH7ACA1FIbNzrU5Do/G7THgY9sfd2lJyZe\nuvKAfROjRp3ddbm5RW+dte3eb2x/yQV/PbZVXmZ+l5rbbghB2AEAKSRZ70/Fjrz7jnnHPnT7\nydslwv4tP2//5yGjzh68+U9zvxtyyV35V0y494y2NbbBFTgVCwCkkoZZK3zkZKzJudeVVpJd\nY0e6isaP/65nr16JEEJI9Oq1yaQJE0qXmz3mg1Hd++Q/ffL+e+xz+Jm3vDGzpu/wdcQOAEgZ\niRDu3H2FKR/NDNe+u3orSUusvJJ5xdUa1XPn9b3+gxBCmz8NublwccjNzVk6PTcvL7lo0ZIQ\nmixbsHjevCVfDv/355efeWr/KUMv3mvPxe9/dEG3am17RcIOAEgZyRAGPF7dlVQkV17Jtm3D\nvl0rWXoV9D7y8sv3CCHkdmjdYnR+1owZM0NoE0KYMX16bn5+k58XzM7Pb5a11wW3ntg/hNCv\n9M2824aHGg07p2IBgFRSK3fFVu/mifweffv27du375adGiZ26LfTmIfuH18aQun4+4eO699/\nu+WX3Hq33RKfjZ1YFkIoHvfxl5n5+dXa8C84YgcApJJauSu25lbV8oi/XXTfblt0GdYxOXlm\nl8uGH9w8hPDxzQOuTr/ikVM3brTv5dfct/NmnYd2bTZrwoLNb31pQM1tOQRhBwCkkLzM8EQN\nt1AIITTLrbl15Wx6wVvfX7DitJ5nPP7I0n+lr3fi09+dWHNbW4mwAwBSRnoi7NSxrgdRj7nG\nDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACAS\nwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAg\nEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4A\nIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIO\nACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLC\nDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACAS\nwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAg\nEsIOACASwg4AIBLCDgAgEsIOACASaxp2H1/cJfHrmh7/4i+WXvjQgIaJRCKRveu9c9ZgY8OP\nb1rJxhKJRL/b56/hawAAiMrvcsRu9tAhzy4OIYSS4Xf/d8rvsUUAgD+e6oVdh7NHJVc2f8hu\nKy018b4hI0rTtjv33G3TK965+54v13Bjm14/6RcbSyaTw09uWq3XAAAQid/hiN24u+8Zlcze\n9YSBF524R1749N6736+o/Y0CAPzh1HrYlb815L7xockBJxzUvMnBJx7ULEx5YMgrpTW+mcUj\nz+6elUi0PnjY9OUnTx92UOtEInPDM99atNyyXz13/Sn7bNm1dZPc7IZrd95kxyMuHzr6h/Kf\nF0jOHzf0sqN33qRDi4bZWXnN2m288zFXPflV4XLrTc4d89DlR/TduHN+09zcJm269ux7yPn3\nvPVd8U8LLBr/5OATdunVtU3zvJxGrdfdeJv9z7j15W+WGwQAQE3LqOX1Fz8/5L/TQ6v/O2Hv\n3BDCHice2e7+fw67+5l/7DqgQY1up8E2gx+/4cPNz3jk+IP/vvHrf+2WGUIo/fymg497dFbD\nbW967PrtGy5bcPpTJ+98+B1fLvnxC4smfTJi0icjXpqTP+fWviGEUD7xwcP6HfPIpJ/is2Tq\np6/d9+lrjz55wYsjrtm2UQiheOTA7fte9/lPS0z/euz0r8e+8b+xaQtf/HPDECq+vH6Xzc57\nb/GP8wsnfjpr4qfvPvnOokkfDuxYySsYOXLkW2+9Vdnrm1awpLJZAABLVS/svr1x88SNK0xp\ncPQLi+77+SK7gseHPDo3dB14fJ+MEEJI2+akYzf65xVPDRk2Z8Cxa63mxj46t1Pi3F9M7fPP\nGSNObR1CyFj/9GH3vtv7wIfPO+DczT74x/bhrXMOOP+dxa0OfPh/SzsvhBCm3H74YXd8WZjV\nfpfT/3bxn3fo3qllmP3NpyMfu/3+Zd+J5JeDDzr2kUk5mx53/WWn7LNl13WapBV8N/aVuy45\n67rXrz36in3G37BVenj9zn99XprW8eB/3HXZgZt3bpG5cPrECaNfuu/m8bmJEEIIo++77b3F\noVX/q++67qjtu66dUzxz0lefvP7QrW81SlT+8l599dVLL720srl9O+Wv5rcLAPjDqd0jdjMf\nuvv5JYnNjz+u549F0/34E7b+2xkv3f2fqcf+tW1Nb671AUOGnT1upxtvPvi4TgOTg275MrnB\nGQ/fc8g6Py3w0b9vfKMwNN3rjvee/nP+siF13LhPx437HL70k4rXbr5pdEnTQx8aMeSAH4/x\nteyy7eGDn2/43dr7Dn344VE3bLVVcunXHTzwtH7dQggh5HTcqGXHjbYdsMJgmux55oV7bxJC\nCCGv/YZbtN9wi73+UtMvGABgOTV9V+zyh+vCpPuGvFaW2ff4o7v8PK3tUSftnlsx8u57v1jd\njf36XbFLD9f9qMH2gx67bodGMx8588xHZzbY5prHru/b6Oe5s9955+sQmh18+lH5lRw6++Lt\nt38IYf7/DmmakZGRkZ6enp6elpaWlpbI3XfowhCmTZxYEkJiuyOO6JiY+O9jj7r6/hdHfTOn\nKLnSWjY55MiNMxc8dOaAS+585t0vZy0u/9VtAQDUrNo8Yjf27ns+Sobw+kmtEyetPO+ze+5+\n7+Ibtqri1OQaylhv/wN6D3zzjZLQaM8zT/7pHGwIIYR58+aFEPLbtq00Z+fMmRNCCBXllbRY\nSUlJCFkNd7l15Ovd/nb9/bf/ZY+LlyTTG7fvuf1uA44987QBGzYKIYTMXleOeLfTNdcMefDc\n/f9WUJ6Wl999m/77H3X66Uf2blHpKz7yyCO33nrryuY+cdHpIybNqPKVAwB/dLUXduUj7rrv\nq8pnT35gyKvXbtUvs/Il1kjxx1cfeP4bJU1btlz0/f9OP/bA7R496OeL05o1axZCmDF1akXY\n8NfbrmnTpiHM3PCSsZ9f2aOqzWTk9zn9tj6n31ZaMHXCJ2M+ePul+/55yQEPPHbVyJEX98oO\nIaQ13/S4Gx477obyxTO//vTjD98Z/tBtg47679D3h3906/KHEJfXqVOnTp06VbbB167K+u0X\nDwD8sdXa250UvXj3Q9NC+o53zPrl6dOC/+ybG74fdvfTNf32H/NfOe2Ayz4q7nDcg+PeuK5P\nw5mPHXfIjV+W/TR77W237RLCvP/984EZK589Xab7Nts0CeGLB//56rxV2mBm47bdt9v7mIG3\nvnzTgPSi0bfe/e6K89MbtF5/y92OOOuGZ+88tmX5V3fd/lLJmr42AIDfUFthN/+xIY/NC5n9\nDj9w7V/ObLTfYXvmhkVPDBk2twY3mfzuvj8dftfEjF4XP3LrHq02/OvDQw7KX/jWwAPOf+un\n9x3Z9JSz++SF+c+ctNWu5/3nzc+nzF2yZO6349546IrDd7387RBCyNj1rLM2yQwThwzoc/g1\n/3t3wtS5hcVL5kz58qNXhw7+y57bnDc8hBCm3XnQNocO/PdT73zy9YyC4tLC+VPGPHX1HSPK\nQ6ioqAghFD9yUu/9z/zHIyPGTJg6r6i0qGDGZy//45/PfR9CcukCAAC1oZZOxc56cMizhSFr\n78MPaP5rsxvufdheDR995OW7//vdcae3W8V1/vrbnYQWJ70+5/a+oWTsNQf95dk5zfr/+7HL\nN88JIYTWh9w9bOS4nW656ZATthn90AGtQwih/ckPPTRm58Pu/PKV64965frl1/J/J4YQQkjr\nftET93/Z/5ih44ZedMjQi1bcVJPjTgshhOSS794d9ui7wwavNJJGW5//lx1CCMniWWOefGrM\nkzevND+n+5ln7JWziq8WAGB11c4Ru2/uv/uNspCz++H7N/v1BfL2PGyfJqHi3bvv/bxGNjj/\n1b8ecOn7Re2OfODBkzv99JoabX/DY4O2aTBj6DGH/nPCstsh1tn3jlFjnrz6+N16dWzRICur\nUat1e+505BVDX7li+2VflN7psIfGjH1y8Cl7b9m1dZOczKxGrTp223yXIy+8/YX3rt8lhBDa\nnvzoe/+77tQB2/dYt3Wj7JzGa3fYcKv9//rPl8e+evZGGSGEnEPuGvPUP885dKde67VpmpPd\ncK1262+2xynXPzF65HXb5tXIywUA+BWJZLKSy82oTy7s0/PaN8eGEAZ1zjm/XXZdDye1dTho\nfl0PIR7f3lvzt7b/YbU40k/jGtOgth+r9AdQ8MrgBU8PXPpvqZBCav1ZsQAA/D6EHQBAJOo4\n7CbfsFWiSlvdMLluRwgAkCocsQMAiEQdX1/a8Zz3kufU7RAAACLhiB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkMup6AMSiovyBKaUTQ9o+\n7bN6+3uhmmY9vuCjT0LjnRptt4PvZXUVhSvnhorMcHrL0Lyux5LyZj9eOPqT0HinnG12SNT1\nWFJd+ec3LZpckNbx2Ebd2tf1WIiKsKOGVJQ/8G3xqyGjbVthV22zHl/w/IOhTUYDYVd9ReHK\nGaE8Nxwp7Kpv1uOFLz0Y1snI2maH9LoeS6or++KmghHT0vv2E3bULL81AAAiIewAACIh7FgT\n8xeWvVWySkuWlpS9uDBZy8NJacWlX75cVroqS1ZUTHmueEFtjye1jSsIkytWackfloSRq/Rt\n/8MqLh//ckXZqixZkfzuuTJ7ZhWS3xR/9vmq/RxcWPrZiDI/MqkGYcfqKy8757PFO3ywcMC3\npV+VV7FYxRPfLun+weLdPyt6tYrF/tiSU66a869dZ1y5y/xRYysq/2menDe84P7e06/f6/th\nD/peVqZkSTj067DBZ+G8uaGKzCguCTdMCl2+CHtPCbN/v9Glmu+uWnTnrvOv2WXJh2OTVXTG\nvOFFD/ae//e9Fj724KoV9R9QsmzEMd/f3mPGTScvmjSr8sXKyyfdMe+mrjNv32PuG1//fqMj\nOsKO1VceNmiW3rKi4onJS7p/sPiMmeU/rLxEctTMwh0+WDhgculXFYktm6VlipFKJEOH7HXb\nhWXddszCr6at/Du06PMlT++5rPxyNspp2zTp92clFofQt2EIJeH6SWHdL8ItC8PKh+TKw7Bp\nYYPPwrlzQ0F66JcXCh0aqUQydMjo3O7Hbjum6JtpKy9R9HnJc3suK7+cjTLbNA32zF9XlGy8\ndXbLnGXdds/VRXMKV1oiOef5gnuWlV+i5Y5ZjUrsmKyxRLKqv8WoLy7s0/PaN8eGEAZ1zjm/\nXXZdDyeEEJYUlf3r26LrZ5XPToamDTIv6pD21OfFb4eMi7tlTPy2eOjiZDKR2KpV9mUdsnbL\nqV9vjNDhoPl1PYSVlJR/fU/B89cumjAlGfIyepzdtPMnc558KqxzWovtSxY+N6RkUXnI65G3\n4yVN+h6QmVu/vpnf3lu/xhNCmLYoXDs9DFkYikPo2jQMbhAOmhbKc8N9LcO/p4f3y0Jaeji4\nVbhk7dCtnt3Y2eLI+vbTuKTim3uKXrq26KspIeSlb3R2bqdPFj3zVMg/rcG2JcUvDilbVB7y\nemTtcElunwPS69n/5g3q21s+VMwp+fCGghdvK/x+UUhvm9Pn6txFF8z7YHr6Frc2afjEgjde\nLS8PibX3arTrZY022yytfhxyKXhl8IKnBy79t1RIIcIuNdTDsFtqSVHZ7d8WXT+rfOby+1Ei\nsU2r7Ms6ZPWvZz/rl6p/YbdUafk39xY8f82i8d+u8P9k3iZ5O13apO/+mfXym1kPw26p6YvC\noBnhroJQtNzEtIxwyNrh0rXDBvUs6Zaqf2G3VGnFxHuLXrqmeMLKe2ZWn0tzd9i/viXdUvUu\n7Jaq+KHkoxsLXry1cPbC5Scn1t670W6XN9q0dz1JuqWEXYqqTzsRKSgvJ+Os9Rt+1iO710+/\nKdPTz+/R6J31s+tn1dVjmenrntjstM9b7bfjT/9bJtY+puVFo9fabUD9rLr6bJ2G4Zau4Z22\nodmPU7JywhPdwkP59bTq6rHMtM4n5p3yeaO9d/x5L2x5TMPzRjfsP6B+Vl39ldYia/Nr1rpw\ndPMebX6clEjvcXfrC59usnn9qjpSl/2IaiksKvvHhEXdPyke89NVdOXlgz9ZuP2E4uFF/sJb\nLaXlE4fMu7X7rCdf/+lSpeTse7+/ZtMfXnqy1DdzNc1YFM78Kmw7Ncz7cUpJURjweThyZhjv\nis/VU1oxaciS27svfOb1n/fC7+9ddP2mi4Y/WV5sz1wdybklH14859recz/56ZrFZPknx826\ndr8FH46p4vYpWHXCjjVUWFT29wmLOn+w+K8zygvzMq/rlr1dCCFkXNwt57C88M6Mol0+WLjd\nhOJXFMlvKy2feNe8W9ebftMJC7+cnb7xxWvtu28IIeSf2uLgE7MS4xY/s/+MS3v/8NIT/9/O\n3cVWXd4BHH/OS98OfUWgBQIUEJnURQTHhUwZwTEVN1iziyUYWEbGlsWFjMmNkmYR3eKSLQ4v\n1OxCuHBmF2Yxm8jWeSGJizGEwJSEly4GMqkSBFuwL5T27EJuRtZSoHraH5/P7XkufvnnSf7f\n//8858i70ei8kLYcT/OOpt93p1n16dWZKZdSqkq7ZqelKb38YWp5Pz3SmY7Iu6sbGPrgDz0v\n3Na180d9R09n79he/fDalFJqenTS9zbnM+9dfP27XU8uudD+Z3l3dcWzF/c/ceap5o93P917\npq5y5a6GZTNSSrllz01euSpz5rXu3UtOPb2ua/8BeccNEnZcu4HBZ49dmPfuZ1s7Bz8py/9s\nYfW/7y5sa8h+fvRvTkPFH++ueWdh+b1lxbc7+1a/e375sf63/WHYMIb+89K55xac+t3m80dO\nZmZvbNhybPqPdxQaq1NKKdtYcd+LTb88NGX1g/mBg5/9pbWz7a5POBgB3gAABeBJREFU/rb3\nkt8eDuPTnstJt7M7TSqkZ29Lh+en1srLny6fmt65I73SlGYNpZdPpZb30/rOdNJNdBjFD1/q\neX5B187NfUdPplkbC48eq9u0o3za5Z2ZX/5i7ROHqlc9mL148OKe1q4n77rQvnfIzvz/egcO\nfp50v+o9ncoW75iy/djU1o1lhUxKKRUWT2r9x/Ttf61bvCh9/Fr37qWnnlrbdfCIjcl1E3Zc\nu2zxvbODH2Wz6+YUDi+btLMpd8uVKzLLmqr2Lat5dU7Zgmzxn2eHemy0YWSGOvqOn0gNq2o3\nHJixbVfNgplXHlmqbCl8Z8/0tvb6r92Z7T3Ue+JMxsUcRiGl9vOpWJ4ea04dt6ctNansfxdk\ncun7M9ORlvSbyanmUtrbkyqcEBtGZqhjoONEsX5V5foD9T/fVTl/5pUrKlvKH95T/3h7Yemd\nmd5DAyfPuJ8MozJzbl/f6Z7c3M0NW483bdpeNaXqihWZKWtqN/1r+tbnq5unFU//vf9czsbk\nuo3PHw4xvuXyv22Z9IOK/L3lIy/LtjYXvj3j0pv9uW86rj6MzOy2KT9dkZm/Ol824rKG+2s3\nHqhe+cZA3RrXcjjlhfSnW1NNdWoesTAqytO2uemHjeloWWr8smabeGa1VW9ekZm/OjvyXaLh\n/spHDlSseGOwdo2uG0Ym/41dU6f1l7csGjHXcrm5P2n4xfrqw/szixZ8WbMRkLDjetTX5O8d\n3cqy8vwDI/ffza6i7CurR7cym529Zhz91c249NXa0a68pZDu+SInmfgqcgtHuzMzs9a4l4wk\nM7+iZZRLa8paVn6hsxCeRywAgCCEHQBAEF6fM0ayuQ1zKr6esks8LNy4xta6h25Ntfe5lmOg\nMrVNT0NlaXKpB4mgsbXqW3bm2MjfvrW2qjvbPLvUgxCNsGOMZHMbmh3rHyONrXUPtZZ6iCgq\nU9uMUs8QxrTWqgfszLGRW7S1blGphyAiz10AAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC\n2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEI\nOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhh\nBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQ+VIPwLV569NLpR5hwutuf6bUI8Tx\nzEelniCQvjftzDEz6K3FDevveKvUI3A9MsVisdQzcHWPr1j8632HSj0FADcjqTCBeKgBAAhC\n2E0Qg4OlngAAGO+csZsYGtZumNf5QqmnAADGNWfsuEntnttV6hEiWPdBqScI4fWWe0o9QgSn\n+vtLPUIQj3V0lHoErp+vYgEAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh\n7AAAghB2AABBCDsAgCCEHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBCDsAgCCE\nHQBAEMIOACAIYQcAEISwAwAIQtgBAAQh7AAAghB2AABBZIrFYqlnAABgDHhjBwAQhLADAAhC\n2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEI\nOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhh\nBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHs\nAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQd\nAEAQwg4AIAhhBwAQhLADAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQhLAD\nAAhC2AEABCHsAACCEHYAAEEIOwCAIIQdAEAQwg4AIAhhBwAQxH8BoW3U+MPbi9MAAAAASUVO\nRK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ME-trait correlation heatmap\n",
    "plotMEtraitCor(MEtraitCor, moduleOrder = 1:length(unique(MEtraitCor$module)),\n",
    "               traitOrder = 1:length(unique(MEtraitCor$trait)),\n",
    "               colColorMargins = c(-2.42, 5.1, 2.7, 4.95),\n",
    "               legend.position = c(1.15, 0.50),\n",
    "               legend.text.size = 6,\n",
    "               legend.title.size = 12,\n",
    "               file = \"ME_Trait_Correlation_Heatmap.pdf\", height = 2.25, width = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3270a51-9bc2-4dc6-8a8b-116707121738",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the GTF annotation file\n",
    "gtf_file <- \"/share/lasallelab/genomes/mm10/mm10.refGene.gtf\"\n",
    "gtf_data <- import(gtf_file)\n",
    "\n",
    "# View\n",
    "head(gtf_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b3876a1-1c45-432c-a81e-02c25993a503",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create GRanges object for regions\n",
    "gr_regions <- GRanges(seqnames = regions$chr,\n",
    "                      ranges = IRanges(start = regions$start, end = regions$end))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38407b6-6e69-4612-88c0-f71085f5e697",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract transcript entries from the GTF data\n",
    "gtf_transcripts <- gtf_data[gtf_data$type == \"transcript\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0407efc2-a275-41f9-9475-419474cdac50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create GRanges object for gene annotations\n",
    "gr_genes <- GRanges(seqnames = seqnames(gtf_transcripts),\n",
    "                    ranges = IRanges(start = start(gtf_transcripts), end = end(gtf_transcripts)),\n",
    "                    gene_name = mcols(gtf_transcripts)$gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72b36843-ec4f-49ec-a563-5fb668e875c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find overlaps between regions and gene annotations\n",
    "overlaps <- findOverlaps(gr_regions, gr_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd511205-dbb4-4248-b22b-16b7576d4e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new column for gene names in the regions data frame\n",
    "regions$gene_name <- NA\n",
    "regions$gene_name[queryHits(overlaps)] <- gr_genes$gene_name[subjectHits(overlaps)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "374e5a50-1a15-4c57-a9bd-f5c860d2ea25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View annotated regions\n",
    "head(regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a71c454-4791-4d4a-a2b5-c81f9b593986",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a list of gene names per module\n",
    "module_genes <- regions %>%\n",
    "  group_by(module) %>%\n",
    "  summarize(unique_genes = list(unique(gene_name[!is.na(gene_name)])), .groups = 'drop')\n",
    "\n",
    "# View\n",
    "head(module_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a5c0dd-d268-46aa-ad4d-994d6e003e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count genes per module\n",
    "module_genes_length <- module_genes %>%\n",
    "  mutate(num_genes = sapply(unique_genes, length)) %>%\n",
    "  select(module, num_genes)\n",
    "\n",
    "# View\n",
    "print(module_genes_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2506908-96af-412c-b0f9-139c41152be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over each module\n",
    "for (module in module_genes$module) {\n",
    "  # Read in gene list\n",
    "  gene_list <- module_genes$unique_genes[[which(module_genes$module == module)]]\n",
    "    \n",
    "  tryCatch({\n",
    "    # Perform the enrichR analysis on the gene list for the current module\n",
    "    enrichr_results <- enrichr(gene_list, c(\"GO_Biological_Process_2023\",\n",
    "                                                     \"GO_Cellular_Component_2023\",\n",
    "                                                     \"GO_Molecular_Function_2023\",\n",
    "                                                     \"KEGG_2019_Mouse\",\n",
    "                                                     \"Panther_2016\",\n",
    "                                                     \"Reactome_2016\",\n",
    "                                                     \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\"))\n",
    "    \n",
    "    # Check if enrichr_results is empty\n",
    "    if (length(enrichr_results) == 0) {\n",
    "      cat(\"No results for module\", module, \"\\n\")\n",
    "      next\n",
    "    }\n",
    "    \n",
    "    # Save Enrichr outputs\n",
    "    wb <- createWorkbook()\n",
    "    \n",
    "    for (i in seq_along(enrichr_results)) {\n",
    "      # Extract the data frame from the list\n",
    "      df <- enrichr_results[[i]]\n",
    "      \n",
    "      # Check if the data frame is empty\n",
    "      if (nrow(df) == 0) {\n",
    "        cat(\"Empty data frame for\", names(enrichr_results)[i], \"in module\", module, \"\\n\")\n",
    "        next\n",
    "      }\n",
    "      \n",
    "      # Define the original sheet name\n",
    "      original_sheet_name <- names(enrichr_results)[i]\n",
    "      \n",
    "      # Modify the sheet name if it's specifically \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\"\n",
    "      sheet_name <- if (original_sheet_name == \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\") {\n",
    "        \"RNAseq_DiseaseGene_DrugSigs_GEO\"\n",
    "      } else {\n",
    "        original_sheet_name\n",
    "      }\n",
    "      \n",
    "      # Add the data frame as a new sheet in the Excel workbook\n",
    "      addWorksheet(wb, sheet_name)\n",
    "      writeData(wb, sheet = sheet_name, x = df)\n",
    "    }\n",
    "    \n",
    "    # Save the Excel workbook\n",
    "    saveWorkbook(wb, paste0(module, \"_enrichr_results.xlsx\"), overwrite = TRUE)\n",
    "    \n",
    "    # Function to plot and save the results\n",
    "    plot_and_save <- function(df, filename, title) {\n",
    "      if (nrow(df) == 0) {\n",
    "        cat(\"Empty data frame for\", title, \"in module\", module, \"\\n\")\n",
    "        return()\n",
    "      }\n",
    "      pdf(filename, height = 7, width = 15)\n",
    "      print(plotEnrich(df, showTerms = 25, numChar = 75, y = \"Count\", orderBy = \"P.value\") + ggtitle(title))\n",
    "      dev.off()\n",
    "    }\n",
    "    \n",
    "    # Plot and save Enrichr results\n",
    "    plot_and_save(enrichr_results$GO_Biological_Process_2023, \n",
    "                  paste0(module, \"_GO_Biological_Process_2023.pdf\"), \n",
    "                  paste(\"GO_Biological_Process_2023 for\", module, \"module\"))\n",
    "    \n",
    "    plot_and_save(enrichr_results$GO_Cellular_Component_2023, \n",
    "                  paste0(module, \"_GO_Cellular_Component_2023.pdf\"), \n",
    "                  paste(\"GO_Cellular_Component_2023 for\", module, \"module\"))\n",
    "    \n",
    "    plot_and_save(enrichr_results$GO_Molecular_Function_2023, \n",
    "                  paste0(module, \"_GO_Molecular_Function_2023.pdf\"), \n",
    "                  paste(\"GO_Molecular_Function_2023 for\", module, \"module\"))\n",
    "    \n",
    "    plot_and_save(enrichr_results$KEGG_2019_Mouse, \n",
    "                  paste0(module, \"_KEGG_2019_Mouse.pdf\"), \n",
    "                  paste(\"KEGG_2019_Mouse for\", module, \"module\"))\n",
    "    \n",
    "    plot_and_save(enrichr_results$Panther_2016, \n",
    "                  paste0(module, \"_Panther_2016.pdf\"), \n",
    "                  paste(\"Panther_2016 for\", module, \"module\"))\n",
    "    \n",
    "    plot_and_save(enrichr_results$Reactome_2016, \n",
    "                  paste0(module, \"_Reactome_2016.pdf\"), \n",
    "                  paste(\"Reactome_2016 for\", module, \"module\"))\n",
    "    \n",
    "    plot_and_save(enrichr_results$`RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO`, \n",
    "                  paste0(module, \"_RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO.pdf\"), \n",
    "                  paste(\"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO for\", module, \"module\"))\n",
    "    \n",
    "  }, error = function(e) {\n",
    "    cat(\"Error occurred for module\", module, \": \", conditionMessage(e), \"\\n\")\n",
    "    # Log the error to a file for further inspection\n",
    "    write(paste(\"Error occurred for module\", module, \": \", conditionMessage(e), \"\\n\"), file = \"error_log.txt\", append = TRUE)\n",
    "    # Continue to the next module\n",
    "    next\n",
    "  })\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1273adf-0884-4f35-a76c-39e24372e56d",
   "metadata": {},
   "source": [
    "## Top 25 Dot Plots (Priotizing Overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91973907-049b-4953-9c9b-e6253e7180ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of GO databases\n",
    "databases <- c(\"GO_Biological_Process_2023\", \"GO_Cellular_Component_2023\", \"GO_Molecular_Function_2023\",\n",
    "               \"KEGG_2019_Mouse\", \"Panther_2016\", \"Reactome_2016\", \"RNAseq_DiseaseGene_DrugSigs_GEO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6fabb01-5178-4e6b-b7a7-a9daaefb4841",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store GO data into data frame\n",
    "\n",
    "# Initialize an empty dataframe\n",
    "all_data <- data.frame()\n",
    "\n",
    "# Iterate over each module and read in the corresponding Excel file\n",
    "for (module in module_genes$module) {\n",
    "  file_path <- glue(\"{module}_enrichr_results.xlsx\")\n",
    "  \n",
    "  for (database in databases) {\n",
    "    try({\n",
    "      # Read the data from the Excel file\n",
    "      df <- read_excel(file_path, sheet = database)\n",
    "      \n",
    "      if (nrow(df) == 0) {\n",
    "        next  # Skip to the next database if no data\n",
    "      }\n",
    "      \n",
    "      # Select the required columns and add module and database information\n",
    "      df <- df %>%\n",
    "        select(Term, Adjusted.P.value, Odds.Ratio) %>%\n",
    "        mutate(Module = module, Database = database)\n",
    "      \n",
    "      # Append to the dataframe\n",
    "      all_data <- bind_rows(all_data, df)\n",
    "      \n",
    "    }, silent = TRUE)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Check if any data was read\n",
    "if (nrow(all_data) == 0) {\n",
    "  stop(\"No data read from any of the Excel files.\")\n",
    "}\n",
    "\n",
    "# Display the combined dataframe\n",
    "head(all_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01818b62-70bb-4fc6-9361-0fd252b9c971",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out rows where the Adjusted.P.value is less than 0.1\n",
    "all_data <- all_data %>%\n",
    "  filter(Adjusted.P.value <= 0.05)\n",
    "\n",
    "# Display the filtered dataframe\n",
    "head(all_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0c6ed7a-8b23-4934-a731-c320df3e4cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the number of total terms and unique terms for each database\n",
    "database_term_counts <- all_data %>%\n",
    "  group_by(Database) %>%\n",
    "  summarise(\n",
    "    Total_Terms = n(),\n",
    "    Unique_Terms = n_distinct(Term)\n",
    "  )\n",
    "\n",
    "# Print the table\n",
    "print(database_term_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d43622e-4083-4276-a389-5ceeca64a99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the number of modules each term appears in for each database\n",
    "term_module_counts <- all_data %>%\n",
    "  group_by(Database, Term) %>%\n",
    "  summarise(ModuleCount = n_distinct(Module), .groups = 'drop')\n",
    "\n",
    "# Rank the terms within each database by the number of modules they appear in\n",
    "ranked_terms <- term_module_counts %>%\n",
    "  arrange(Database, desc(ModuleCount)) %>%\n",
    "  group_by(Database) %>%\n",
    "  slice_head(n = 25) %>%\n",
    "  ungroup()\n",
    "\n",
    "# Merge with the original data to filter the top 25 terms per database\n",
    "filtered_data_top_25 <- all_data %>%\n",
    "  semi_join(ranked_terms, by = c(\"Database\", \"Term\"))\n",
    "\n",
    "# Print the filtered data\n",
    "head(filtered_data_top_25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b15982b-8cf5-4081-a884-6244c812fb17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the number of total terms and unique terms for each database\n",
    "filtered_database_term_counts <- filtered_data_top_25 %>%\n",
    "  group_by(Database) %>%\n",
    "  summarise(\n",
    "    Total_Terms = n(),\n",
    "    Unique_Terms = n_distinct(Term)\n",
    "  )\n",
    "\n",
    "# Print the table\n",
    "print(filtered_database_term_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e5f450a-1eb5-445c-a3fb-e8c4e41c8c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plots for each database\n",
    "for (database in unique(filtered_data_top_25$Database)) {\n",
    "  database_filtered_data <- filtered_data_top_25 %>%\n",
    "    filter(Database == database)\n",
    "  \n",
    "  if (nrow(database_filtered_data) > 0) {\n",
    "    dot_plot <- ggplot(database_filtered_data, aes(x = Module, y = Term, size = Odds.Ratio, fill = Adjusted.P.value)) +\n",
    "      geom_point(shape = 21) +\n",
    "      scale_fill_viridis() +\n",
    "      xlab('') + ylab('') +\n",
    "      labs(\n",
    "        title = 'Top Enrichr Terms Across Modules',\n",
    "        subtitle = glue('{database}')\n",
    "      ) +\n",
    "      theme(\n",
    "        panel.background = element_rect(fill = \"white\", color = NA),\n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor = element_blank(),\n",
    "        panel.border = element_blank(),\n",
    "        axis.line = element_line(color = \"black\"),\n",
    "        plot.background = element_rect(fill = \"white\", color = NA),\n",
    "        axis.text.x = element_text(angle = 90, hjust = 1)\n",
    "      )\n",
    "    \n",
    "    # Save the dot plot with the database name in the filename\n",
    "    ggsave(filename = glue(\"top_25_dot_plot_{database}.pdf\"), plot = dot_plot, height = 7, width = 15)\n",
    "  } else {\n",
    "    cat(glue(\"No data available for {database}. Skipping...\\n\"))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e52d00c-b9d2-45cb-8381-a3180a614acc",
   "metadata": {},
   "source": [
    "## Top 25 Per Module Dot Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "412c9680-3edc-42fe-8df3-c688ca6013cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of GO databases\n",
    "databases <- c(\"GO_Biological_Process_2023\", \"GO_Cellular_Component_2023\", \"GO_Molecular_Function_2023\",\n",
    "               \"KEGG_2019_Mouse\", \"Panther_2016\", \"Reactome_2016\", \"RNAseq_DiseaseGene_DrugSigs_GEO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce4d438a-ea83-4e06-847c-e35d2d2cef55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store GO data into data frame\n",
    "\n",
    "# Initialize an empty dataframe\n",
    "all_data <- data.frame()\n",
    "\n",
    "# Iterate over each module and read in the corresponding Excel file\n",
    "for (module in modules_of_interest) {\n",
    "  file_path <- glue(\"{module}_enrichr_results.xlsx\")\n",
    "  \n",
    "  for (database in databases) {\n",
    "    try({\n",
    "      # Read the data from the Excel file\n",
    "      df <- read_excel(file_path, sheet = database)\n",
    "      \n",
    "      if (nrow(df) == 0) {\n",
    "        next  # Skip to the next database if no data\n",
    "      }\n",
    "      \n",
    "      # Select the required columns and add module and database information\n",
    "      df <- df %>%\n",
    "        select(Term, Adjusted.P.value, Odds.Ratio) %>%\n",
    "        mutate(Module = module, Database = database)\n",
    "      \n",
    "      # Append to the dataframe\n",
    "      all_data <- bind_rows(all_data, df)\n",
    "      \n",
    "    }, silent = TRUE)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Check if any data was read\n",
    "if (nrow(all_data) == 0) {\n",
    "  stop(\"No data read from any of the Excel files.\")\n",
    "}\n",
    "\n",
    "# Display the combined dataframe\n",
    "head(all_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc0cbd02-f51a-4640-8c06-73f770ec8d05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out rows where the Adjusted.P.value is less than 0.1\n",
    "all_data <- all_data %>%\n",
    "  filter(Adjusted.P.value <= 0.05)\n",
    "\n",
    "# Display the filtered dataframe\n",
    "head(all_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3914472e-2597-41ca-8ad9-a0e2d7fa622a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the number of total terms and unique terms for each database\n",
    "database_term_counts <- all_data %>%\n",
    "  group_by(Database) %>%\n",
    "  summarise(\n",
    "    Total_Terms = n(),\n",
    "    Unique_Terms = n_distinct(Term)\n",
    "  )\n",
    "\n",
    "# Print the table\n",
    "print(database_term_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1143215-621a-4014-95ea-33e30befd0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter to keep only the top 10 most significant per module and database\n",
    "top_25 <- all_data %>%\n",
    "  group_by(Module, Database) %>%\n",
    "  slice_min(order_by = Adjusted.P.value, n = 25, with_ties = FALSE) %>%\n",
    "  ungroup()\n",
    "\n",
    "# View \n",
    "head(top_25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0744a2c9-0603-4264-8ae5-007e5ee72dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the number of total terms and unique terms for each database\n",
    "database_term_counts <- top_25 %>%\n",
    "  group_by(Database) %>%\n",
    "  summarise(\n",
    "    Total_Terms = n(),\n",
    "    Unique_Terms = n_distinct(Term)\n",
    "  )\n",
    "\n",
    "# Print the table\n",
    "print(database_term_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6befd2c0-fbb8-4324-8e0c-3f7df14ce7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plots for each database\n",
    "for (database in unique(top_25$Database)) {\n",
    "  # Filter for the current database\n",
    "  current_data <- top_25 %>%\n",
    "    filter(Database == database)\n",
    "  \n",
    "  if (nrow(current_data) > 0) {\n",
    "    dot_plot <- ggplot(current_data, aes(x = Module, y = Term, size = Odds.Ratio, fill = Adjusted.P.value)) +\n",
    "      geom_point(shape = 21) +\n",
    "      scale_fill_viridis() +\n",
    "      xlab('') + ylab('') +\n",
    "      labs(\n",
    "        title = 'Top 25 Enrichr Terms Per Module',\n",
    "        subtitle = glue('{database}')\n",
    "      ) +\n",
    "      theme(\n",
    "        panel.background = element_rect(fill = \"white\", color = NA),\n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor = element_blank(),\n",
    "        panel.border = element_blank(),\n",
    "        axis.line = element_line(color = \"black\"),\n",
    "        plot.background = element_rect(fill = \"white\", color = NA),\n",
    "        axis.text.x = element_text(angle = 90, hjust = 1)\n",
    "      )\n",
    "    \n",
    "    # Save the dot plot with the database name in the filename\n",
    "    ggsave(filename = glue(\"top_25_per_module_dot_plot_{database}.pdf\"), plot = dot_plot, height = 12, width = 15)\n",
    "  } else {\n",
    "    cat(glue(\"No data available for {database}. Skipping...\\n\"))\n",
    "  }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
